---
title: "Polyclonality and percent altered"
author: "Aidan Flynn"
date: "09/07/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r data_load}
library(ggplot2)
library(dplyr)
library(tidyr)
library(patchwork)
source("C:/Users/AFFLY/OneDrive - The University of Melbourne/ResearchData/RADIO/A5/Scripts/dataloaders/load_All_Data.R")

dataloader_cna_segs()
dataloader_sv_gridsslinx()
dataloader_somatic_variants(quickload = F)  
```

## Percent Genome Altered

```{r percent_altered}




no_xy_genome_size <- sum(chr_offsets$chr_size[!(chr_offsets$chromosome %in% c("chrX","chrY"))])

genome_altered_stats <- A5_seg.keep.merged %>% 
  filter(minorAlleleCopyNumber < 0.70 | majorAlleleCopyNumber > 1.2, !(chromosome %in% c("chrX","chrY"))) %>%  
  mutate(segsize=end-start) %>% 
  group_by(A5_ID) %>% 
  summarise(TotalAltered=sum(segsize), 
            pcnt_altered=(TotalAltered/no_xy_genome_size)*100) 

genome_altered_stats <- genome_altered_stats %>% mutate(A5_ID=gsub("T0","",A5_ID)) %>% left_join(a5_anno %>%  dplyr::select(A5_ID, is_primary_or_met,tumour_metastasised, polyclonalProportion)) %>% 
  mutate(ClinicalCourse=case_when(
    is_primary_or_met == "Primary" & tumour_metastasised == "No" ~ "Non-malignant Primary",
    is_primary_or_met == "Primary" & tumour_metastasised == "Yes" ~ "Malignant Primary",
    is_primary_or_met == "Metastatic" & tumour_metastasised == "Yes" ~ "Metastasis",
    TRUE ~ "Other"
  ), ClinicalCourse=factor(as.character(ClinicalCourse), levels=c("Non-malignant Primary", "Malignant Primary", "Metastasis", "Other"))) 


ggplot(genome_altered_stats, 
       aes(x=ClinicalCourse, 
           y=pcnt_altered)) + 
  geom_boxplot(outlier.alpha = 0) + 
  geom_jitter(width=0.2) + 
  theme_bw() + theme(axis.text.x = element_text(angle=90, vjust = 0.5,hjust=1))
```

## Polyclonal Proportion (purple metric)

```{r polyclonal }


ggplot(a5_anno %>%  
  mutate(ClinicalCourse=case_when(
    is_primary_or_met == "Primary" & tumour_metastasised == "No" ~ "Non-malignant Primary",
    is_primary_or_met == "Primary" & tumour_metastasised == "Yes" ~ "Malignant Primary",
    is_primary_or_met == "Metastatic" & tumour_metastasised == "Yes" ~ "Metastasis",
    TRUE ~ "Other"
  ), ClinicalCourse=factor(as.character(ClinicalCourse), levels=c("Non-malignant Primary", "Malignant Primary", "Metastasis", "Other")))
    , 
       aes(x=ClinicalCourse, 
           y=as.numeric(polyclonalProportion))) + 
  geom_boxplot(outlier.alpha = 0) + 
  geom_jitter(width=0.2) +
  theme_bw() + theme(axis.text.x = element_text(angle=90, vjust = 0.5,hjust=1)) + ylab("Polyclonal Proportion")
```
```{r polyclonal_vs_pcntaltered, eval=F, include=F }
ggplot(genome_altered_stats %>% 
         filter(is_primary_or_met %in% c("Primary","Metastatic"), 
                tumour_metastasised %in% c("Yes","No")), 
       aes(color=interaction(is_primary_or_met,tumour_metastasised), 
           x=pcnt_altered,
           y=as.numeric(polyclonalProportion))) + 
  geom_point() + 
  theme_bw() + theme(axis.text.x = element_text(angle=90, vjust = 0.5,hjust=1))
```
# SV Size classes
```{r sv_distribution, fig.height=12, fig.width=8, eval=FALSE}

sv_sizes <- A5_gridss.keep %>% 
  dplyr::select(A5_ID, chrom1, start1, end1, chrom2, end2, start2, resolvedClusterType) %>% 
  mutate(sv_size=ifelse(chrom1==chrom2, start2-start1, 0), 
         size_class=cut(sv_size,breaks = c(-1,1,10^4,10^10), labels = c("TRA","<10k",">10k"))) %>% 
  distinct()

sv_sizes$Chromothripsis = FALSE
for (cr in 1:nrow(chromothripsis_regions))
{
  #BreadEnd1
  sv_sizes <- sv_sizes %>%
    mutate(Chromothripsis = ifelse(
             (
               A5_ID == chromothripsis_regions$A5_ID[[cr]] | A5_ID == gsub("T0", "", chromothripsis_regions$A5_ID[[cr]])
             ) &
               (chrom1 == chromothripsis_regions$chromosome[[cr]] &
                  ((start1 >= chromothripsis_regions$start[[cr]]) &
                     (end1 <= chromothripsis_regions$end[[cr]])
                  )),
             TRUE,
             Chromothripsis
           ))
  #BreadEnd2
  sv_sizes <- sv_sizes %>%
    mutate(Chromothripsis = ifelse(!is.na(chrom2) &
             (
               A5_ID == chromothripsis_regions$A5_ID[[cr]] | A5_ID == gsub("T0", "", chromothripsis_regions$A5_ID[[cr]])
             ) &
               (chrom2 == chromothripsis_regions$chromosome[[cr]] &
                  ((start2 >= chromothripsis_regions$start[[cr]]) &
                     (end2 <= chromothripsis_regions$end[[cr]])
                  )),
             TRUE,
             Chromothripsis
           ))
}

sv_sizes.summary.with_chrtps <- sv_sizes %>% filter(!is.na(size_class)) %>% mutate(A5_ID=gsub("T0","", A5_ID)) %>% left_join(a5_anno %>% dplyr::select(A5_ID, is_primary_or_met,tumour_metastasised)) %>% 
  group_by(A5_ID, is_primary_or_met,tumour_metastasised, size_class) %>% summarise(sv_count=n()) %>% mutate(Chromothripsis="Including Chromothripsis")

sv_sizes.summary.no_chrtps <- sv_sizes %>% filter(!is.na(size_class), !Chromothripsis) %>% mutate(A5_ID=gsub("T0","", A5_ID)) %>% left_join(a5_anno %>% dplyr::select(A5_ID, is_primary_or_met,tumour_metastasised)) %>% 
  group_by(A5_ID, is_primary_or_met,tumour_metastasised, size_class) %>% summarise(sv_count=n()) %>% mutate(Chromothripsis="Excluding Chromothripsis")

sv_sizes.summary <- bind_rows(sv_sizes.summary.with_chrtps, sv_sizes.summary.no_chrtps) %>% mutate(sv_count_log2=log2(sv_count)) %>% 
  mutate(ClinicalCourse=case_when(
    is_primary_or_met == "Primary" & tumour_metastasised == "No" ~ "Non-malignant Primary",
    is_primary_or_met == "Primary" & tumour_metastasised == "Yes" ~ "Malignant Primary",
    is_primary_or_met == "Metastatic" & tumour_metastasised == "Yes" ~ "Metastasis",
    TRUE ~ "Other"
  ), ClinicalCourse=factor(as.character(ClinicalCourse), levels=c("Non-malignant Primary", "Malignant Primary", "Metastasis", "Other")))


ggplot(sv_sizes.summary %>% 
         filter(ClinicalCourse != "Other"), 
       aes(x=ClinicalCourse, 
           y=sv_count,
           color=size_class)) + 
  geom_boxplot(outlier.alpha = 0) + 
  geom_point(position = position_jitterdodge(jitter.width = 0.2)) + 
  theme_bw() + theme(axis.text.x = element_text(angle=90, vjust = 0.5,hjust=1)) + 
  facet_wrap("Chromothripsis", scales = "free") +
ggplot(sv_sizes.summary %>% 
         filter(ClinicalCourse != "Other"), 
       aes(x=ClinicalCourse, 
           y=sv_count_log2,
           color=size_class)) + 
  geom_boxplot(outlier.alpha = 0) + 
  geom_point(position = position_jitterdodge(jitter.width = 0.2)) + 
  theme_bw() + theme(axis.text.x = element_text(angle=90, vjust = 0.5,hjust=1)) + 
  facet_wrap("Chromothripsis", scales = "free") +
  plot_layout(nrow=2)

```

```{r hartwig_sv}

complex_sv <- A5_gridss %>% filter(resolvedClusterType=="COMPLEX") %>% 
  dplyr::select(A5_ID, clusterId, clusterCount, resolvedClusterType) %>% 
  distinct() 

inversion_sv <- A5_gridss %>% filter(resolvedClusterType %in% c("INV", "FB_INV_PAIR", "RECIP_INV", "RECIP_INV_DEL_DUP", "RECIP_INV_DUPS")) %>% dplyr::select(A5_ID, clusterId, clusterCount, resolvedClusterType) %>% distinct()

translocation_sv <- A5_gridss %>% filter(resolvedClusterType %in% c("RECIP_TRANS","RECIP_TRANS_DEL_DUP", "RECIP_TRANS_DUPS", "UNBAL_TRANS", "UNBAL_TRANS_T")) %>% dplyr::select(A5_ID, clusterId, clusterCount, resolvedClusterType) %>% distinct()

complex.summary <- bind_rows(complex_sv, inversion_sv, translocation_sv) %>% 
  mutate(class_label=cut(clusterCount,breaks = c(0,20,10^6), labels = c("Complex SVs (<20 breakpoints)","Complex SVs (>=20 breakpoints)"))) %>% group_by(A5_ID, class_label) %>% dplyr::count() 

del_dup_sv <-  A5_gridss %>% filter(resolvedClusterType %in% c("DEL","DUP")) %>% dplyr::select(A5_ID, chrom1, start1, end1, chrom2, end2, start2, resolvedClusterType, clusterId, clusterCount) %>% distinct() %>% mutate(sv_size=ifelse(chrom1==chrom2, start2-start1, 0), 
         size_class=cut(sv_size,breaks = c(0,10^4,10^10), labels = c("<10k",">10k"))) %>% 
  mutate(class_label=case_when(
    resolvedClusterType=="DEL" & size_class== "<10k" ~ "Small Deletions (<10kb)",
    resolvedClusterType=="DEL" & size_class== ">10k" ~ "Large Deletions (>=10kb)",
    resolvedClusterType=="DUP" & size_class== "<10k" ~ "Small Duplications (<10kb)",
    resolvedClusterType=="DUP" & size_class== ">10k" ~ "Large Duplications (>=10kb)"))

del_dup.summary <- del_dup_sv %>% filter(!is.na(size_class)) %>% group_by(A5_ID, class_label) %>% dplyr::count()

sv.summary <-  bind_rows(del_dup.summary, complex.summary)
  
sv.summary <- sv.summary %>% full_join(crossing(A5_ID=unique(sv.summary$A5_ID), class_label=unique(sv.summary$class_label))) %>% 
  mutate(n=replace_na(n,0))
  
sv.summary <- sv.summary %>% mutate(A5_ID=gsub("T0","", A5_ID)) %>% left_join(a5_anno %>% dplyr::select(A5_ID, is_primary_or_met,tumour_metastasised)) %>% 
  mutate(ClinicalCourse=case_when(
    is_primary_or_met == "Primary" & tumour_metastasised == "No" ~ "Non-malignant Primary",
    is_primary_or_met == "Primary" & tumour_metastasised == "Yes" ~ "Malignant Primary",
    is_primary_or_met == "Metastatic" & tumour_metastasised == "Yes" ~ "Metastasis",
    TRUE ~ "Other"
  ), ClinicalCourse=factor(as.character(ClinicalCourse), levels=c("Non-malignant Primary", "Malignant Primary", "Metastasis", "Other"))) %>% 
  dplyr::select(-is_primary_or_met, -tumour_metastasised)

```
# SV size plot hartwig preprint style
```{r hartwig_sv_size}
sv.summary %>% group_by(class_label, ClinicalCourse) %>% arrange(class_label, ClinicalCourse, n) %>% mutate(row_idx=row_number()) %>% 
  ggplot(aes(x=row_idx, y=n, color=ClinicalCourse)) + geom_point() + facet_wrap("class_label", nrow=6, scale="free")
```
# SV size boxplot 
```{r sv_size_boxplot}
ggplot(sv.summary, aes(x=ClinicalCourse, y=n)) +
 geom_boxplot(outlier.alpha = 0) + 
  geom_point(position = position_jitter(width = 0.2, height=0)) + 
  theme_bw() + theme(axis.text.x = element_text(angle=90, vjust = 0.5,hjust=1)) + 
  facet_wrap("class_label", scales = "free")

```
# Annotate variants with purple subclonality score
```{r hartwig_snv_clonality}

subcl_scores <- read.delim("C:/ResearchData/RADIO/A5/Data/WGS/hg38/somatic_variants/purple_variant_subclonality_scores.tsv")
subcl_scores <- subcl_scores %>%  mutate(A5_ID=gsub("T0","", A5_ID))

A5_vcfs <- A5_vcfs %>% left_join(subcl_scores, by=c("A5_ID", "seqnames"="CHROM", "start"="POS", "REF", "ALT"))

subcl.summary <- A5_vcfs %>% filter(!is.na(SUBCL)) %>% mutate(SUBCL=ifelse(SUBCL==".", 0, as.numeric(SUBCL)), SUBCL_binary=SUBCL>0.8) %>% group_by(A5_ID, SUBCL_binary) %>% dplyr::count() %>%  group_by(A5_ID) %>% mutate(prop_subcl=n/sum(n)) %>% left_join(a5_anno %>%  dplyr::select(A5_ID, PublicationID, is_primary_or_met,tumour_metastasised, polyclonalProportion)) %>% 
  mutate(ClinicalCourse=case_when(
    is_primary_or_met == "Primary" & tumour_metastasised == "No" ~ "Non-malignant Primary",
    is_primary_or_met == "Primary" & tumour_metastasised == "Yes" ~ "Malignant Primary",
    is_primary_or_met == "Metastatic" & tumour_metastasised == "Yes" ~ "Metastasis",
    TRUE ~ "Other"
  ), ClinicalCourse=factor(as.character(ClinicalCourse), levels=c("Non-malignant Primary", "Malignant Primary", "Metastasis", "Other"))) 

```
# Sanity check of VAF vs Subclonal score
Does sub-clonality score sensibly reflect VAF? Variants with no score set as -0.5 for plot
```{r subclonality_score_vs_vaf}

A5_vcfs %>% filter(A5_ID %in% c("E019-1", "E120-1", "E159-1", "E159-3")) %>% mutate(SUBCL=ifelse(SUBCL==".", -0.5, as.numeric(SUBCL))) %>%  ggplot(aes(x=SUBCL, y=as.numeric(AF))) + geom_jitter(height=0, width=0.05, alpha=0.5) + geom_vline(xintercept = 0.8, linetype=2, color="red") + facet_wrap("A5_ID") + theme_bw()

```

## Hartwig style subclonal proportion plot
```{r primary_vs_met_subclonprop}

plot.data1 <- subcl.summary %>% filter(SUBCL_binary, ClinicalCourse %in% c("Non-malignant Primary","Malignant Primary","Metastasis")) %>%  group_by(is_primary_or_met) %>%  summarise(mean_prop_subclonal=mean(prop_subcl)) %>% mutate(patientID="Average_All") %>% 
  pivot_wider(id_cols = patientID, names_from = is_primary_or_met, values_from=mean_prop_subclonal)


has_pri_met_pair <- a5_anno %>%  filter(`Patient ID` %in% intersect(a5_anno %>% filter(is_primary_or_met == "Metastatic") %>% pull(`Patient ID`),
a5_anno %>% filter(is_primary_or_met == "Primary") %>% pull(`Patient ID`))) %>%  pull(A5_ID)

plot.data2 <- subcl.summary %>% mutate(subcode=gsub(".+-","", PublicationID)) %>% 
  filter(A5_ID %in% has_pri_met_pair, SUBCL_binary==T, ClinicalCourse %in% c("Malignant Primary","Metastasis")) %>% 
  mutate(patientID=gsub("-.","",A5_ID)) %>% 
  pivot_wider(id_cols=c(patientID), names_from = subcode, values_from=prop_subcl) %>% dplyr::select(-P2) %>% pivot_longer(cols = c(M1,M2), names_to="discard", values_to="M") %>% filter(!is.na(M)) %>% dplyr::select(-discard) %>% dplyr::rename(Primary=P1, Metastatic=M)

plot.data <- bind_rows(plot.data1, plot.data2)

ggplot(plot.data, aes(x=Primary, y=Metastatic, color=patientID, group==patientID, size=patientID)) + 
  geom_point() + 
  geom_abline(intercept = 0, slope=1) + 
  geom_line(size=0.3) +
  scale_size_manual(values=c(3,1.3,1.3,1.3,1.3,1.3,1.3)) +
  coord_cartesian(xlim=c(0,1), ylim=c(0,1)) + theme_bw() + theme(aspect.ratio = 1)

```
## Proportion subclonal boxplot
```{r prop_subclonal_boxplot}
subcl.summary %>%  filter(SUBCL_binary==T) %>%  ggplot(aes(x=ClinicalCourse, y=prop_subcl)) + geom_boxplot(outlier.alpha = 0) + 
    geom_point(position = position_jitter(width = 0.2, height=0)) + 
    theme_bw() + theme(axis.text.x = element_text(angle=90, vjust = 0.5,hjust=1)) 
```
## Proportion subclonal vs purples proportion polyclonal score
```{r polyclon_vs_subclone}
subcl.summary %>%  filter(SUBCL_binary==T) %>% ggplot(aes(x=prop_subcl, y=as.numeric(polyclonalProportion))) + geom_point() + theme_bw()
```