---
title: "snRNA_pseudobulk_de"
author: "Blake Bowen"
date: "20/12/2021"
output: 
  html_document:
    code_folding: show
    fig_crop: false
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = '/data/gpfs/projects/punim0648/Projects/Blake/A5_R_project/')

rm(list=ls())

library(tidyverse)
library(patchwork)
library(limma)
library(edgeR)
library(Homo.sapiens)
library(scales)
library(RColorBrewer)
library(ggsci)
library(ggrepel)

```

Differential expression and Gene-set enrichment Analysis for Pseudobulked snRNA-seq

TODO: 
[] remove adrenocortical signature genes from the differential gene expression
[] make volcano plots to visualise the results
[] remove the 3' chemistry samples from the comparison?

# Tumour and chromaffin cell differential gene-expression

```{r load-data, include=FALSE}

# read in the pseudo-bulked scRNA
counts <- read_csv("Data/snRNA-seq/processed/snRNA_pseudobulk.csv") %>%
  dplyr::rename("gene" = `...1`) %>% 
  dplyr::select(starts_with(c("gene", "Chromaffin cells", "Tumor"))) %>% # interested only in the chromaffin cell / neoplastic populations
  dplyr::select(-`Tumor_E240-1`, -`Tumor_E243-1`) %>% # these correspond to mis-labelled normal chromaffin cells, therefore removing them
  # name columns according to A5 ID
  dplyr::rename(!!!setNames(colnames(.), str_remove(colnames(.), "Tumor_|Chromaffin cells_"))) 

A5_ids <- counts %>% 
  dplyr::select(-gene) %>% 
  colnames() %>%
  sort()

counts <- counts %>%
  dplyr::select(all_of(sort(colnames(.)))) %>%  # reorder alphanumerically
  column_to_rownames(var = "gene") %>% 
  as.matrix()

```

```{r organise-metadata}

# TODO: determine genes overexpressed within adrenocortical cell populations so these may be removed from the tumour-vs-normal comparison to mitigate ambient rna contamination

# Read in the clinical data ----
# TODO: put in the latest clinical data 
clinical_data <- read_csv("Data/A5_full_clinical_and_genomic_version_2.csv") %>%
  filter(`A5 ID` %in% A5_ids) %>% 
  mutate(`Metastatic state of tumour sample` = replace(`Metastatic state of tumour sample`, `Metastatic state of tumour sample` == "Metastatic", "Metastasis")) %>%
  # Add in patient info for MDS later on
  mutate(Patient = gsub("-.*", "", `A5 ID`)) %>%
  mutate(Patient_plot = replace(Patient,!(duplicated(Patient)| duplicated(Patient,fromLast = T)),NA))

# get available metadata for the pheo-atlas paper samples
singlecell_metadata <- read_tsv("Data/metadata_sc_samples.tsv") 

# harmonize the important single cell metadata fields with A5 metadata  
singlecell_metadata <- singlecell_metadata %>%
  dplyr::rename("A5 ID" = Sample) %>% 
  filter(`A5 ID` %in% c("P018-PGL1", "P018-PGL3", "E240", "E243")) %>% 
  dplyr::select(`A5 ID`, Gender, Location, Malignancy) %>% 
  mutate(Gender = recode(Gender,"M" = "male")) %>% 
  mutate(`Metastatic state of tumour sample` = if_else(
    `A5 ID` %in% c("E240", "E243"),
    "Normal", "Primary")) %>%  # these have malignancy = "benign" so they must all be primaries 
  mutate(`Tumour metastasised` = if_else(
    `A5 ID` %in% c("E240", "E243"),
    "Normal", "No")) %>% 
  dplyr::select(-Malignancy, -Location) %>% 
  mutate(`A5 ID` = case_when(`A5 ID` == "E240" ~ "E240-1",
                             `A5 ID` == "E243" ~ "E243-1",
                             TRUE ~ `A5 ID`)) %>% 
  mutate(`Patient ID` = str_extract(string = `A5 ID`,
                                    pattern = "[^-]*"))

metadata <- clinical_data %>%
  bind_rows(singlecell_metadata) %>% 
  arrange(`A5 ID`) %>% 
  # make a column describing the groups to compare
  mutate(genotype_groups = case_when(`ATRX mutant` == "Yes" ~ "ATRX",
                                     `TERT mutant` == "Yes" ~ "TERT",
                                     `A5 ID` %in% c("E240-1", "E243-1") ~ "Normal_chromaffin", 
                                     TRUE ~ "TERT_and_ATRX_wildtype_tumor")) %>% 
  mutate(chemistry = if_else(`A5 ID` %in% c("E140-1",
                                        "E143-1",
                                        "E171-1",
                                        "E225-1"),
                         "3prime", "5prime"))

# make column names that work better with r
compatiblenames <- setNames(colnames(metadata), make.names(colnames(metadata)))
metadata <- metadata %>%
  dplyr::rename(!!!compatiblenames)

```

```{r rle-plot}

# TODO: make ggplot rle plots 


```

```{r limma-preprocessing, cache=TRUE}

# Preprocess RNA-seq ----
counts_dge <- DGEList(counts)
# filter genes with low expression in all/nearly all samples
keep.exprs <- filterByExpr(counts, group = metadata$genotype_groups)
counts_dge <- counts_dge[keep.exprs,, keep.lib.sizes = FALSE]
# Apply TMM normalisation to the DGElist
counts_dge <- calcNormFactors(counts_dge)

# Convert the TMM counts to log2 CPMs
lcpm <- cpm(counts_dge, log = T)

```

MDS plots help to visualise the overall similarity/difference between gene expression profiles of the cohort. They are useful for determining which clinical factors that may be responsible for this variation. 
- samples appear to cluster by chemistry
- samples cluster depending on whether they are normal or tumor
- TERT mutant samples appear to cluster, with the exception of E143-1
- MDS also indicates co-clustering of P018-PGL1 and P018-PGL3 which are synchronous primaries from the same patient

```{r mds-plots, fig.width=14, fig.height=8, cache=TRUE}

mds <- plotMDS(lcpm)
mds_tbl <- tibble(leading_logFC_dim1 = mds$x,
                  leading_logFC_dim2 = mds$y,
                  `A5.ID` = colnames(mds$distance.matrix),
                  Dim = "1+2") %>% 
  left_join(metadata) %>%
  # Select key columns
  dplyr::select(A5.ID,
                leading_logFC_dim1,
                leading_logFC_dim2,
                genotype_groups,
                Gender,
                Metastatic.state.of.tumour.sample,
                Tumour.metastasised,
                chemistry,
                Patient.ID) 
# Plot some possible and expected sources of variation
mds_ggplot <- ggplot(data = mds_tbl, aes(x = leading_logFC_dim1, y = leading_logFC_dim2)) 

mds1 <- mds_ggplot +
  geom_point(aes(colour = genotype_groups))+
  geom_text_repel(aes(label = A5.ID))
mds2 <- mds_ggplot +
  geom_point(aes(colour = Gender))
mds3 <- mds_ggplot +
  geom_point(aes(colour = Metastatic.state.of.tumour.sample))
mds4 <- mds_ggplot +
  geom_point(aes(colour = Tumour.metastasised))
mds5 <- mds_ggplot +
  geom_point(aes(colour = chemistry))
mds6 <- mds_ggplot +
  geom_point(aes(colour = Patient.ID))
mds1 + mds2 + mds3 + mds4 + mds5 + mds6

```

I want to compare TERT mutant PCPG with ATRX mutant and Wild-Type PCPG and Vice-Versa (same comparison groups as the bulk).

I also want to compare all tumour vs all normal samples to get a tumour-specific gene list. 

```{r differential-expression, cache=TRUE}

# make a design matrix 
genotype_groups <- metadata$genotype_groups
sex <- metadata$Gender
chemistry <- metadata$chemistry
design_mat <- model.matrix(~0 + genotype_groups + sex + chemistry)

# comparison of TERT vs ATRX, TERT vs rest and ATRX vs rest
contr.matrix <- makeContrasts(
  atrx_vs_normal = genotype_groupsATRX - genotype_groupsNormal_chromaffin, 
  tert_vs_normal = genotype_groupsTERT - genotype_groupsNormal_chromaffin, 
  atrx_vs_tert = genotype_groupsATRX - genotype_groupsTERT, 
  atrx_vs_wt = genotype_groupsATRX - (genotype_groupsTERT + genotype_groupsTERT_and_ATRX_wildtype_tumor + genotype_groupsNormal_chromaffin)/3,
  tert_vs_wt = genotype_groupsTERT - (genotype_groupsATRX + genotype_groupsTERT_and_ATRX_wildtype_tumor + genotype_groupsNormal_chromaffin)/3,
  levels = colnames(design_mat))

par(mfrow=c(1,2))

# voom transform
v <- voom(counts_dge, design_mat, plot=TRUE)

# differential expression analysis
vfit <- lmFit(v, design_mat)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit, main="Final model: Mean-variance trend")

sum.fit <- decideTests(efit)
summary(sum.fit)

```

```{r toptables, cache=TRUE}

# top ATRX signature genes (ATRX vs TERT comparison)
atrx_vs_normal_tt <- topTable(efit, coef = "atrx_vs_normal", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))
atrx_vs_normal_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/atrx_vs_normal_dge_hg38.tsv")
# make top table figure
atrx_vs_normal_tt %>%
  slice_max(order_by = abs(logFC), n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top ATRX differentially expressed genes (ATRX-mutant PCPG vs Normal Chromaffin cells comparison)")

# Get the top TERT signature genes 
tert_vs_normal_tt <- topTable(efit, coef = "tert_vs_normal", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))
tert_vs_normal_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/tert_vs_normal_dge_hg38.tsv")
# make top table figure
tert_vs_normal_tt %>%
  slice_max(order_by = abs(logFC), n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top TERT differentially expressed genes (TERT-mutant PCPG vs Normal Chromaffin cells comparison)")

# top ATRX signature genes (ATRX vs TERT comparison)
atrx_vs_tert_tt <- topTable(efit, coef = "atrx_vs_tert", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))
atrx_vs_tert_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/atrx_vs_tert_dge_hg38.tsv")
# make top table figure
atrx_vs_tert_tt %>%
  slice_max(order_by = abs(logFC), n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top ATRX differentially expressed genes (ATRX-mutant PCPG vs TERT-mutant PCPG)")

# get top ATRX signature genes (vs all wt)
atrx_vs_wt_tt <- topTable(efit, coef = "atrx_vs_wt", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))
atrx_vs_wt_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/atrx_vs_wt_dge_hg38.tsv")
# make top table figure
atrx_vs_wt_tt %>%
  slice_max(order_by = abs(logFC), n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top ATRX differentially expressed genes (ATRX-mutant PCPG vs all ATRX-wildtype PCPG and Normal Chromaffin cells)")

# Get the top TERT signature genes 
tert_vs_wt_tt <- topTable(efit, coef = "tert_vs_wt", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))
tert_vs_wt_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/tert_vs_wt_dge_hg38.tsv")
# make top table figure
tert_vs_wt_tt %>%
  slice_max(order_by = abs(logFC), n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top TERT differentially expressed genes (PCPG-mutant PCPG vs all TERT-wildtype PCPG and Normal Chromaffin cells)")

```

Looking at the top differentially expressed genes for TERT vs normal and ATRX vs normal comparisons, it is clear that there are a large number of overlapping genes - the venn diagram shows 180 overlapping genes. 

```{r, venn-diagram}

par(mfrow=c(1,1))
vennDiagram(sum.fit[,1:3], circle.col=c("turquoise", "salmon"))

```

These overlapping genes are predominantly adrenocortical cell signature genes, which and are likely artifacts of ambient RNA contamination from adrenocortical cells, specific to the adrenal medulla samples.

# Normal cell differential gene expression

Next, I compare adrenocortical cells with other normal cell populations to establish a list of adrenocortical-specific genes.

```{r, adrenocortical-signature, cache=TRUE}

# organise the normal counts data ----
# read in the pseudo-bulked scRNA
norm_counts <- read_csv("Data/snRNA-seq/processed/snRNA_pseudobulk.csv") %>%
  dplyr::rename("gene" = `...1`) %>% 
  dplyr::select(-starts_with("Tumor")) %>% # interested only in the normal cells
  dplyr::rename(!!!setNames(colnames(.), str_replace(colnames(.), " ", "_"))) # format the colnames

norm_samples <- norm_counts %>% 
  dplyr::select(-gene) %>% 
  colnames() %>%
  str_extract(pattern = "[^_]*$")

norm_metadata <- tibble(name = colnames(norm_counts)[colnames(norm_counts)!="gene"],
                        A5.ID = norm_samples) %>% 
  arrange(name) %>%  
  mutate(cell_type = str_remove(name, pattern = "_[^_]*$")) %>% 
  left_join(metadata %>% dplyr::select(A5.ID, Gender))

norm_counts <- norm_counts %>%
  dplyr::select(all_of(sort(colnames(.)))) %>%  # reorder alphanumerically
  column_to_rownames(var = "gene") %>% 
  as.matrix()

norm_dge <- DGEList(norm_counts)
# filter genes with low expression in all/nearly all samples
keep.exprs <- filterByExpr(norm_counts, group = norm_metadata$cell_type)
norm_dge <- norm_dge[keep.exprs,, keep.lib.sizes = FALSE]
# Apply TMM normalisation to the DGElist
norm_dge <- calcNormFactors(norm_dge)

# Convert the TMM counts to log2 CPMs
norm_lcpm <- cpm(norm_dge, log = T)

# design matrix
cell_type <- norm_metadata$cell_type
norm_sex <- norm_metadata$Gender
norm_design_mat <- model.matrix(~0 + cell_type + norm_sex)

# setup contrast between Adrenocorticall cells and all of the other normal cell types including tumor microenvironment and healthy adrenal tissue
norm_contr_mat <- makeContrasts(
  adrenocortical = cell_typeAdrenocortical_cells - (cell_typeChromaffin_cells + cell_typeEndothelial_cells + cell_typeFibroblasts + cell_typeLymphocytes + cell_typeMyeloid_cells + cell_typeSCLCs)/6, levels = colnames(norm_design_mat))

# voom transform
par(mfrow=c(1,2))
norm_v <- voom(norm_dge, norm_design_mat, plot=TRUE)

norm_vfit <- lmFit(norm_v, norm_design_mat)
norm_vfit <- contrasts.fit(norm_vfit, contrasts=norm_contr_mat)
norm_efit <- eBayes(norm_vfit)
plotSA(norm_efit, main="Final model: Mean-variance trend")

norm_sum.fit <- decideTests(norm_efit)
summary(norm_sum.fit)

# Get the top adrenocortical signature genes 
adrenocortical_tt <- topTable(norm_efit, coef = "adrenocortical", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))
adrenocortical_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/adrenocortical_dge.tsv")
# make top table figure
adrenocortical_tt %>%
  dplyr::select(-gene_symbol) %>% 
  slice_max(order_by = abs(logFC), n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top Adrenocortical cell differentially expressed genes (Adrenocortical cells vs all other normal cells)")

```

Applying a logFC threshold of > 1 and a adjusted p-value theshold of < 0.05 to the adrenocortical DEG list gives _ genes that are upregulated in the adrenocortical cell compartment. These can be removed from the tumor-normal comparisons to remove likely ambient-rna-associated genes.

```{r, filter-tobtables, cache=TRUE}

# get the genes to remove
ac_sig_genes <- adrenocortical_tt %>%
  filter(logFC > 1, adj.P.Val < 0.05) %>% 
  pull(Gene)
ac_sig_genes

atrx_vs_normal_tt_filtered <- atrx_vs_normal_tt %>%
  filter(!Gene %in% ac_sig_genes) %>% # remove ambient genes if downregulated in tumor
 # filter(abs(logFC) > 1 & adj.P.Val > 0.05) %>%  # remove non-significant DEGs and those that are below the logFC cutoff
  arrange(desc(abs(logFC)))

atrx_vs_normal_tt_filtered %>%
  slice_max(order_by = abs(logFC), n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top ATRX differentially expressed genes (ATRX-mutant PCPG vs Normal Chromaffin cells comparison)")

tert_vs_normal_tt_filtered <- tert_vs_normal_tt %>%
  filter(!Gene %in% ac_sig_genes) %>% # remove ambient genes if downregulated in tumor
  # filter(abs(logFC) > 1 & adj.P.Val > 0.05) %>%  # remove non-significant DEGs and those that are below the logFC cutoff
  arrange(desc(abs(logFC)))

tert_vs_normal_tt_filtered %>%
  slice_max(order_by = abs(logFC), n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top TERT differentially expressed genes (TERT-mutant PCPG vs Normal Chromaffin cells comparison)")

```


```{r, volcano-plots, fig.height=8, cache=TRUE}

plot_volcano <- function(tt, n_label){
  #' plots a volcano plot
  #' @param tt a toptable produced by limma  
  #' @param n_label the number of genes to label

  
  # get the top n over and underexpressed genes from the toptable
  genes_to_label_high <- tt %>%
    filter(adj.P.Val < 0.05) %>% 
    slice_max(logFC, n = n_label%/%2) %>% 
    pull(Gene)
  genes_to_label_low <- tt %>%
    filter(adj.P.Val < 0.05) %>% 
    slice_min(logFC, n = n_label%/%2) %>% 
    pull(Gene)
  genes_to_label <- c(genes_to_label_high, genes_to_label_low)
  
  tt_plot <- tt %>% #flag the most significant genes to label
    mutate(significant = if_else(adj.P.Val < 0.05, "Adj. P Value < 0.05", "not significant")) %>% 
    mutate(label = if_else(Gene %in% genes_to_label, TRUE, FALSE))
  
  volcano <- ggplot(tt_plot, aes(logFC, -log10(adj.P.Val))) +
    geom_point(aes(col=significant)) +
    scale_color_manual(values=c("red", "black")) +
    geom_text_repel(
      data = . %>% filter(label == TRUE),
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.1, "lines"),
      segment.color = "grey",
      max.overlaps = Inf,
      size = 3,
      aes(label=gene_symbol)) + ggtitle("")
  
  return(volcano)
   
}

tert_vs_normal_volcano <- plot_volcano(tt = tert_vs_normal_tt_filtered, n_label = 10) +
  labs(title = "TERT vs normal chromaffin DEGs")

atrx_vs_normal_volcano <- plot_volcano(tt = atrx_vs_normal_tt_filtered, n_label = 10) +
  labs(title = "ATRX vs normal chromaffin DEGs")

tert_vs_normal_volcano + atrx_vs_normal_volcano + plot_layout(ncol=1)

```

```{r environment}
sessionInfo()
```

