---
title: "A5 GSVA"
author: "Aidan Flynn"
date: "02/08/2022"
output: 
  html_document:
    code_folding: show
    fig_crop: false
editor_options: 
  chunk_output_type: console
---



```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = '/data/cephfs/punim0010/projects/flynna/A5/WTS/',cache = TRUE,cache.lazy = FALSE)

```

```{r libs, include=FALSE}
library(tidyverse)
library(ComplexHeatmap)
library(limma)
library(edgeR)
#library(Homo.sapiens)
#library(ggrepel)
library(patchwork)
#library(circlize)
library(googlesheets4)
#library(umap)
#library(clusterProfiler)
#library(biomaRt)
#library(org.Hs.eg.db)
library(GSVA)
source("scripts/dotplot_functions_A5.R")

ColorPalette <- c(LightBlue1="#cadae8ff",LightBlue2="#abc4e2ff",LightBlue3="#8aa9d6ff",
DarkBlue1="#7884baff",DarkBlue2="#606d9fff",DarkBlue3="#4c5a88ff",
Purple1="#765b92ff",Purple2="#9370a5ff",Purple3="#b280a9ff",Purple4="#cc85b1ff",
LightRed1="#ee7474ff",LightRed2="#e2696aff",LightRed3="#de5353ff",
DarkRed1="#c25858ff",DarkRed2="#c04241ff",DarkOrange1="#c65a44ff",
DarkOrange2="#eb7b54ff",LightOrange1="#f18d73ff",LightOrange2="#f5a697ff",
LightBrown1="#f5b9b0ff",LightBrown2="#d6a097ff",DarkBrown1="#c17963ff",DarkBrown2="#96665aff",
DarkGreen1="#637b62ff",DarkGreen2="#6ea668ff",LightGreen1="#98ca8cff",LightGreen2="#e2eab5ff",
Yellow1="#fbf2adff",Yellow2="#fef8c6ff",Yellow3="#f4e764ff",Salmon="#f2c5a7ff",LightSalmon="#f2dec4ff",
LemonGrey1="#f7f0e2ff",LemonWhite1="#fefbe5ff",LemonWhite2="#f7f4dcff",LemonGrey2="#efecdcff",
LightGrey1="#e2e0d9ff",LightGrey2="#d6d6d4ff",DarkGrey1="#b4b4b4ff",DarkGrey2="#9a9999ff")

theme_set(theme_bw())
# can add more to the theme later
```

```{r gene_sets}
Hs.H <- msigdbr::msigdbr(species = "Homo sapiens", category = "H")
Hs.C8 <- msigdbr::msigdbr(species = "Homo sapiens", category = "C8")

# put the genesets into list form
Hs.H.list <- lapply(split(Hs.H, f = Hs.H$gs_name), function(w) w$gene_symbol)

#Janskey

signature_genes_jansky <- read_csv("/data/gpfs/projects/punim0648/Projects/Blake/single_cell_R_project/Data/signature_genes/jansky_sympathoadrenal_genes.csv", skip = 1)

jansky.genesets.list <- list(
  JANSKY_late_chromaffin_genes = signature_genes_jansky %>% filter(cluster == "late Chromaffin cells") %>% pull(gene),
  JANSKY_early_chromaffin_genes = signature_genes_jansky %>% filter(cluster == "Chromaffin cells") %>% pull(gene),
  JANSKY_connecting_chromaffin_genes = signature_genes_jansky %>% filter(cluster == "connecting Progenitor cells")  %>% pull(gene),
  JANSKY_bridge_genes = signature_genes_jansky %>% filter(cluster == "Bridge") %>% pull(gene),
  JANSKY_early_neuroblast_genes = signature_genes_jansky %>% filter(cluster == "Neuroblasts") %>% pull(gene),
  JANSKY_late_neuroblast_genes = signature_genes_jansky %>% filter(cluster == "late Neuroblasts") %>% pull(gene),
  JANSKY_cycling_SCP_genes = signature_genes_jansky %>% filter(cluster == "cycling SCPs") %>% pull(gene),
  JANSKY_cycling_neuroblast_genes = signature_genes_jansky %>% filter(cluster == "cycling Neuroblasts") %>% pull(gene),
  JANSKY_early_SCP_genes = signature_genes_jansky %>% filter(cluster == "SCPs") %>% pull(gene),
  JANSKY_late_SCP_genes = signature_genes_jansky %>% filter(cluster == "late SCPs") %>% pull(gene))

pseudobulk_dge_all <- read_csv("/data/gpfs/projects/punim0648/Projects/Blake/single_cell_R_project/Data/Pseudobulk all DGE.csv")

# filter out genes below significance threshold cutoff of 0.05 adj p value
# filter out genes with log fold change below 3 
pseudobulk_dge <- pseudobulk_dge_all %>% 
  filter(grepl("_normal" , Contrast)) %>%
  mutate(Contrast = Contrast %>% str_replace_all("\\.", "_") %>%
           str_replace("Sustentacular_cells", "SCLCs") %>% 
           str_remove("_normal")) %>% 
  filter(adj.P.Val < 0.05) %>% 
  filter(logFC > 3)

# chose logFC cutoff to have a good number of genes in each set (200-400 genes)
# this can be tweaked further if necessary
zethoven.celltypes.list <- lapply(split(pseudobulk_dge,
                                        f = pseudobulk_dge$Contrast),
                                  function(y) y$Gene)
names(zethoven.celltypes.list) <- paste0("ZETHOVEN_", names(zethoven.celltypes.list))

all_genesets <- c(Hs.H.list, jansky.genesets.list, zethoven.celltypes.list)

```

```{r load-data, include=FALSE, message=FALSE}
# Read in the clinical data ----
# TODO: put in the latest clinical data when this becomes available

gs4_auth("aidan.flynn@umccr-radio-lab.page")
a5_anno <- read_sheet("1hnXdXI29KvvuLxsaTBID1-EbE7mTSk6bG05HcSFfhgo", col_types = "c")
a5_anno <- a5_anno %>% dplyr::rename(`A5_ID`=`A5 ID`)
a5_anno <- a5_anno %>% filter(!is.na(`A5_ID`), Exclude=="N")


feature_sheet <- as_sheets_id("1IOczHu91crprHvjxTKDRpqLfpeuYLaGwNXZVsWmXIKY")
features <- read_sheet(ss = feature_sheet,sheet = "SampleFeatures", col_types = "c")

# Read in the htseq counts ----
htseq_outs <- "Data/bulk RNA-seq/htseq/truseq/gene"

# read in the counts for each sample
counts_df <- list.files(htseq_outs, full.names = T) %>% 
  map(function(x){
    sample_name <- sub(pattern = ".gene.counts", replacement = "", basename(x))
    read_tsv(x,col_types = "ccd") %>%
      dplyr::rename(!!sample_name := count)
  }) %>% 
  purrr::reduce(left_join, by = c("gene_id", "gene_symbol")) %>% 
  mutate(gene = paste0(gene_id, "_", gene_symbol)) %>% 
  # remove the rows without gene symbols (these look like they are just QC rows from htseq)
  dplyr::filter(!is.na(gene_symbol)) %>% 
  dplyr::select(-gene_id, -gene_symbol) %>% 
  column_to_rownames(var = "gene")

```
```{r organise-cohort, include=FALSE}

# reorder clinical data in same order as counts
a5_anno <- data.frame(A5_ID = colnames(counts_df),
                            stringsAsFactors = F,
                            check.names = F)%>%
  left_join(a5_anno) %>% 
  filter(!is.na(`Patient ID`)) # removes  E181-1 (No WGS), E191-1 (No WGS), and low purity samples (E227-1, E154-1)

# write_csv(counts %>% rownames_to_column(var = 'gene'), "data/tpm_counts_ensg_with_symbols.csv")
# write_csv(a5_anno, "data/a5_anno_formatted.csv")

# make a column describing the groups to compare
a5_anno <- a5_anno %>%
  mutate(genotype_groups = case_when(`Assumed driver of metastasis` == "ATRX" ~ "ATRX",
                                     `Assumed driver of metastasis` == "TERT" ~ "TERT", 
                                     TRUE ~ "TERT_and_ATRX_wildtype"))

```
Head and neck PGLs derive from glomus cells and are thus are quite different from the abdominothoracic PGLs and phaeochromoctyomas (which arise from chromaffin cells). As I am mainly interested in the TERT and ATRX mutant tumours, I have removed the head and neck tumours from the cohort to simplify the differential expression analysis.


Other samples to form core analysis set:
E124-1 - NF1 Case
E144-1 - low quality
E145-1 - VHL case

```{r limma-preprocessing}

# Remove samples that were not included in the WGS cohort due to issues with genotype etc.
# and Remove samples that are head and neck tumours 
counts_df <- counts_df[,a5_anno$A5_ID]
counts_df_list <- list()
counts_df_list[["All"]] <- counts_df
counts_df_list[["NoHN"]] <- counts_df[,!(colnames(counts_df) %in% a5_anno$A5_ID[a5_anno$`is_head_and_neck` == "H_N"])]
counts_df_list[["NoHN_Core"]] <- counts_df[,!(colnames(counts_df) %in% c(a5_anno$A5_ID[a5_anno$`is_head_and_neck` == "H_N"],"E124-1", "E144-1", "E145-1", "E154-1"))]

# Preprocess RNA-seq ----
counts_dge_list <- lapply(counts_df_list, DGEList)

# filter genes with low expression in all/nearly all samples
keep.exprs_list <- lapply(counts_dge_list, filterByExpr, group = a5_anno$genotype_groups)
counts_dge_list <- lapply(names(counts_dge_list), 
                          function (dataset_name)  { 
                            counts_dge_list[[dataset_name]][keep.exprs_list[[dataset_name]],, keep.lib.sizes = FALSE] 
                            }
                          )
names(counts_dge_list) <- names(counts_df_list)
# Apply TMM normalisation to the DGElist
counts_dge_list <- lapply(counts_dge_list, calcNormFactors)

# Convert the TMM counts to log2 CPMs

lcpm_list <- lapply(counts_dge_list, cpm, log = T)

# lapply(names(lcpm_list), function (dataset_name, dataset_list) { 
#   base_dir= "results/differential_gene_expression/bulk_RNA_seq"
#   write.csv(dataset_list[[dataset_name]], paste0(base_dir, 
#                                                 "/bulk_counts_hg38_tmm_normalized_log2cpm_",
#                                                 dataset_name,
#                                                 ".csv")
#   )}, dataset_list=lcpm_list)
```
```{r gsva}

gsva_res_list <- list()

gsva_res_list <- lapply(lcpm_list, function (lcpm_df) {
  #strip ENGID to keep only HGNC symbol
  rownames(lcpm_df) <- make.unique(gsub("ENSG[0-9.]+?_","", rownames(lcpm_df), perl = T))
  gsva(lcpm_df, all_genesets, min.sz=10, max.sz=500, method="gsva", kcdf="Gaussian")})


```
```{r gene_set_score_plots}
plot.data <- gsva_res_list$All %>% data.frame(check.names = F) %>% tibble::rownames_to_column("Pathway") %>% 
  filter(Pathway %in% c("E2F_TARGETS","G2M_CHECKPOINT","MITOTIC_SPINDLE","JANSKY_cycling_neuroblast_genes",
"ZETHOVEN_SCLCs","JANSKY_early_SCP_genes","JANSKY_late_SCP_genes",
"HALLMARK_MYOGENESIS",
"ZETHOVEN_T_NK_cells","HALLMARK_IL2_STAT5_SIGNALING",
"ZETHOVEN_Fibroblasts",
"ZETHOVEN_Endothelial_cells") | grepl("chromaffin", Pathway))

plot.data <- plot.data %>% 
  pivot_longer(cols = -Pathway, names_to = "A5_ID", values_to = "Pathway_score") %>% 
  left_join(a5_anno %>%  select(A5_ID, `Patient ID`, is_primary_or_met, tumour_metastasised, Primary_Location_Simplified) %>% 
  mutate(ClinicalCourse=case_when(
    is_primary_or_met == "Primary" & tumour_metastasised == "No" ~ "Non-malignant Primary",
    is_primary_or_met == "Primary" & tumour_metastasised == "Yes" ~ "Malignant Primary",
    is_primary_or_met == "Primary" & tumour_metastasised == "Short follow up" ~ "Short follow up",
    is_primary_or_met == "Metastatic" & tumour_metastasised == "Yes" ~ "Metastasis",
    is_primary_or_met == "Recurrent" & tumour_metastasised == "Yes" ~ "Recurrance w/ Metastasis",
    is_primary_or_met == "Recurrent" & tumour_metastasised == "No" ~ "Recurrance w/o Metastasis",
    TRUE ~ "Other"
  ), ClinicalCourse=ifelse(Primary_Location_Simplified=="Head_neck", paste("Head/Neck", ClinicalCourse, sep=" - "), ClinicalCourse), 
  ClinicalCourse=factor(as.character(ClinicalCourse), levels=c("Non-malignant Primary", "Malignant Primary", "Metastasis",  
                                                               "Other", "Head/Neck - Non-malignant Primary",
                                                               "Head/Neck - Malignant Primary", "Head/Neck - Other",  
                                                               "Recurrance w/ Metastasis", "Recurrance w/o Metastasis", "Short follow up", "Head/Neck - Short follow up"))))  %>% 
  mutate(Primary_Location_Simplified=gsub("_[Ll]eft|_[Rr]ight|_abdominal","",Primary_Location_Simplified))

plot.data <- plot.data %>% left_join(features %>% filter(Gene %in% c("TERT","ATRX"))) %>% mutate(Gene=replace_na(Gene,"None"), Gene=factor(Gene, levels=c("None","ATRX","TERT"))) %>% 
  rename(Mutation_Detected=Gene) %>% 
  arrange(Pathway, ClinicalCourse, `Patient ID`)  

pre_jitter <- position_jitter(width = 0.2, height = 0, seed=10)
pdf(file = "/data/gpfs/projects/punim0648/flynna/A5/WTS/results/plots/box_jitter_plot/gsva_pathway_scores.pdf", width = 14, height=12, onefile = T)
ggplot(plot.data, aes(x=ClinicalCourse, y=Pathway_score,  group=`Patient ID`)) + 
  geom_boxplot(aes(group=NULL), outlier.alpha = 0) + 
  geom_point(aes(color=Mutation_Detected),alpha=0.7, position = pre_jitter) + 
  geom_path(position = pre_jitter, linetype=2, alpha=0.2) +
  facet_wrap("Pathway") + 
  theme(axis.text.x = element_text(angle=90, vjust = 0.5, hjust = 1)) + 
  scale_color_manual(values=c("None"=ColorPalette[["DarkGrey1"]], "TERT" =ColorPalette[["DarkBlue2"]], "ATRX"=ColorPalette[["DarkOrange1"]])) + 
  ggtitle("Colored by mutation, lines link patient samples")


ggplot(plot.data, aes(x=ClinicalCourse, y=Pathway_score, group=`Patient ID`)) + 
  geom_boxplot(aes(group=NULL), outlier.alpha = 0) + 
  geom_point(aes(color=Primary_Location_Simplified),alpha=1, position = pre_jitter) + 
  #geom_path(aes(color=NULL), position = pre_jitter, linetype=2, alpha=0.3) +
  facet_wrap("Pathway") + 
  theme(axis.text.x = element_text(angle=90, vjust = 0.5, hjust = 1)) + 
  scale_color_manual(values=c("Extraadrenal"=ColorPalette[["DarkGreen1"]], "Extraadrenal_thoracic" =ColorPalette[["LightGreen1"]], "Extraadrenal_bladder"=ColorPalette[["DarkBlue1"]], "Adrenal"=ColorPalette[["DarkRed1"]],   "Unspecified"  =ColorPalette[["DarkGrey1"]], "Head_neck"=ColorPalette[["Purple4"]])) + 
  ggtitle("Colored by Primary Location")

ggplot(plot.data, aes(x=ClinicalCourse, y=Pathway_score, group=`Patient ID`)) + 
  geom_boxplot(aes(group=NULL), outlier.alpha = 0) + 
  geom_point(aes(color=Primary_Location_Simplified, shape=Mutation_Detected),alpha=1, position = pre_jitter) + 
  geom_path(position = pre_jitter, linetype=2, alpha=0.3) +
  facet_wrap("Pathway") + 
  theme(axis.text.x = element_text(angle=90, vjust = 0.5, hjust = 1)) + 
  scale_color_manual(values=c("Extraadrenal"=ColorPalette[["DarkGreen1"]], "Extraadrenal_thoracic" =ColorPalette[["LightGreen1"]], "Extraadrenal_bladder"=ColorPalette[["DarkBlue1"]], "Adrenal"=ColorPalette[["DarkRed1"]],   "Unspecified"  =ColorPalette[["DarkGrey1"]], "Head_neck"=ColorPalette[["Purple4"]])) + 
  ggtitle("Colored by Primary Location, Shape by mutation, lines link patient samples")

dev.off()
```
###UMAPS
```{r UMAPS}

set.seed(42)
umap_config <- umap.defaults
umap_config$n_neighbors=10
umap_config$spread=3
umap_list <- lapply(lcpm_list, function (lcpm, umc) { umap(t(lcpm), config = umc) }, umc=umap_config) 

annotate_umap <- function (umap_result, annotation)
{
  umap_layout <- data.frame(umap_result$layout)
  return(
  tibble(UMAP1 = umap_layout$X1,
                  UMAP2 = umap_layout$X2,
                  A5_ID = rownames(umap_layout),
                  Dim = "1+2") %>% 
  left_join(annotation) %>%
  # Select key columns
  dplyr::select(A5_ID,
                UMAP1,
                UMAP2,
                Gender,
                genotype_groups,
                is_primary_or_met,
                tumour_metastasised,
                TERT_or_ATRX,
                chr_14_miRNA_outgroup, 
                `is_head_and_neck`,
                `TERT mutant`,
                `ATRX mutant`,
                #Run,
                Dim,
                Patient_plot,
                Patient,
                `Year of birth`)
  )
}


umap_annotated_list <- lapply(umap_list,annotate_umap, annotation=a5_anno)

write.table(umap_annotated_list$All, "results/plots/umap/umap_coord_AllSample_seed42_nn10_hg38_tmm_log2cpm.tsv", sep="\t", row.names=F)
write.table(umap_annotated_list$NoHN, "results/plots/umap/umap_coord_NoHdNk_seed42_nn10_hg38_tmm_log2cpm.tsv", sep="\t", row.names=F)
write.table(umap_annotated_list$NoHN_Core, "results/plots/umap/umap_coord_NoHdNk_CoreSamples_seed42_nn10_hg38_tmm_log2cpm.tsv", sep="\t", row.names=F)

colour_by <- c("genotype_groups", "is_primary_or_met", "Tumour metastasised",
              "TERT mutant", "ATRX mutant", "is_head_and_neck", "Gender")

umap_plot_list <- lapply(umap_annotated_list, function (umap_tbl) {
  umap_ggplot <- ggplot(data = umap_tbl, aes(x = UMAP1, y = UMAP2))
  
  plot_list <- lapply(colour_by, function (colour_variable, plot_prototype) { 
  plot_prototype + geom_point(aes(colour = !!sym(colour_variable))) }, plot_prototype=umap_ggplot )
  
  plot_list[[length(plot_list)+1]] <-  umap_ggplot + geom_text(aes(label=A5_ID))
  
  return(plot_list)
} )


```
```{r mds-plots, fig.width=18, fig.height=12}
#pdf(file = "results/plots/umap/umap_bulkcounts_hg38_AllSamples.pdf", height = 30, width=15)
Reduce(f = "+", x=umap_plot_list[["All"]] ) + plot_layout(ncol=2) 
#dev.off()
```

```{r mds-plots-nohn, fig.width=18, fig.height=12}
#pdf(file = "results/plots/umap/umap_bulkcounts_hg38_NoHdNk.pdf", height = 30, width=15)
Reduce(f = "+", x=umap_plot_list[["NoHN"]] ) + plot_layout(ncol=2) 
#dev.off()
```

```{r mds-plots-nohn, fig.width=18, fig.height=12}
pdf(file = "results/plots/umap/umap_bulkcounts_hg38_NoHdNk_CoreSamples.pdf", height = 30, width=15)
Reduce(f = "+", x=umap_plot_list[["NoHN_Core"]] ) + plot_layout(ncol=2) 
dev.off()
```

Make a design matrix with genotype groups (TERT-mutant, ATRX-mutant and TERT and ATRX wildtype).

Setup the contrasts to be TERT vs both other groups, and ATRX vs both other groups.

To account for sample-specific effects, I used the patient ID as a blocking variable (as suggested here: <https://stat.ethz.ch/pipermail/bioconductor/2013-July/054087.html>)
```{r differential-expression}

a5_anno.core <-  a5_anno[!(a5_anno$A5_ID %in% c(a5_anno$A5_ID[a5_anno$`is_head_and_neck` == "H_N"],"E124-1", "E144-1", "E145-1", "E154-1")),]



# make a design matrix 
genotype_groups <- a5_anno.core$genotype_groups
sex <- a5_anno.core$Gender
design_mat <- model.matrix(~0 + genotype_groups + sex)

# comparison of TERT vs rest and ATRX vs rest 
contr.matrix <- makeContrasts(
  atrx_vs_wt = genotype_groupsATRX - (genotype_groupsTERT_and_ATRX_wildtype),
  tert_vs_wt = genotype_groupsTERT - (genotype_groupsTERT_and_ATRX_wildtype),
  atrx_vs_tert = genotype_groupsATRX - (genotype_groupsTERT),
  levels = colnames(design_mat))

par(mfrow=c(1,2))

# voom transform
counts_dge <- counts_dge_list[["NoHN_Core"]]
v <- voom(counts_dge, design_mat, plot=TRUE)

# estimate correlation between samples from the same patient  
dupcor <- duplicateCorrelation(v, design_mat, block=a5_anno.core$`Patient ID`)

# differential expression analysis, keeping track of samples that are correlated as they come from the same patient
vfit <- lmFit(v, design_mat, block=a5_anno.core$`Patient ID`, correlation=dupcor$consensus)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit, main="Final model: Mean-variance trend")

sum.fit <- decideTests(efit)
summary(sum.fit)

# get top ATRX signature genes 
atrx_tt <- topTable(efit, coef = "atrx_vs_wt", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))

# atrx_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/atrx_vs_wt_dge_NoHdNk_coreSamples_hg38.tsv")

# Get the top TERT signature genes 
tert_tt <- topTable(efit, coef = "tert_vs_wt", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))

# tert_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/tert_vs_wt_dge_NoHdNk_coreSamples_hg38.tsv")

tert_vs_atrx_tt <- topTable(efit, coef = "atrx_vs_tert", number = Inf, sort.by = "P") %>% 
  rownames_to_column(var = "Gene") %>% 
  mutate(gene_symbol = if_else(grepl(x = Gene, pattern = "_"), 
                             str_extract(Gene, pattern = '[^_]*$'),
                             Gene))

# atrx_vs_tert_tt %>% write_tsv("results/differential_gene_expression/bulk_RNA_seq/tert_vs_atrx_tt_dge_NoHdNk_coreSamples_hg38.tsv")


```

Look at the overlap between the TERT and ATRX signature gene sets. The majority of genes are different.

```{r venn-diagram}
vennDiagram(sum.fit[,1:2], circle.col=c("turquoise", "salmon"))
```

Take a look at some of the top 20 signature genes for each signature.

```{r toptables}

atrx_tt %>%
  slice_min(order_by = adj.P.Val, n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top ATRX differentially expressed genes")

tert_tt %>%
  slice_min(order_by = adj.P.Val, n = 20) %>%
  column_to_rownames(var="Gene") %>%
  knitr::kable(caption = "Top TERT differentially expressed genes")

```

```{r volcano-plots, fig.height=12, fig.width=9}

plot_volcano <- function(tt, n_label){
  #' plots a volcano plot
  #' @param tt a toptable produced by limma  
  #' @param n_label the number of genes to label

  genes_to_label <- tt %>%
  slice_min(adj.P.Val, n = n_label) %>% 
  pull(Gene)

  tt_plot <- tt %>% #flag the most significant genes to label
    mutate(significant = if_else(adj.P.Val < 0.05, "Adj. P Value < 0.05", "not significant")) %>% 
    mutate(label = if_else(Gene %in% genes_to_label, TRUE, FALSE))
  
  volcano <- ggplot(tt_plot, aes(logFC, -log10(adj.P.Val))) +
    geom_point(aes(col=significant)) +
    scale_color_manual(values=c("red", "black")) +
    geom_text_repel(
      data = . %>% filter(label == TRUE),
      box.padding = unit(0.5, "lines"),
      point.padding = unit(0.1, "lines"),
      segment.color = "grey",
      max.overlaps = Inf,
      size = 3,
      aes(label=gene_symbol))
  
  return(volcano)
   
}

tert <- plot_volcano(tert_tt, 20) + ggtitle("TERT Mutant Vs WT")
atrx <- plot_volcano(atrx_tt, 20) + ggtitle("ARTX Mutant Vs WT")
tert_vs_atrx <- plot_volcano(tert_vs_atrx_tt, 20) + ggtitle("TERT Vs ARTX")

tert + atrx + tert_vs_atrx + plot_layout(ncol=1)

```

Visualise the bulk expression signatures as heatmaps, alongside the snRNA-seq data to see which cell types each gene is expressed in.

```{r heatmaps, fig.width=12, fig.height=20}

# ----
# organise the genes to plot 
# ----

tert_tt <- tert_tt %>% mutate(genotype = "TERT")
atrx_tt <- atrx_tt %>% mutate(genotype = "ATRX")

degs_bulk_all <- tert_tt %>% bind_rows(atrx_tt)

# get the top 20 genes that are up and down in each genotype
top_degs_up <- degs_bulk_all %>% 
  group_by(genotype) %>% 
  slice_max(order_by = logFC, n = 20) %>% 
  ungroup()
top_degs_down <- degs_bulk_all %>% 
  group_by(genotype) %>% 
  slice_min(order_by = logFC, n = 20) %>% 
  ungroup()

genes_plot_df <- bind_rows(top_degs_up, top_degs_down) %>%
  mutate(direction = ifelse(logFC > 0, "up", "down")) %>% 
  arrange(genotype, direction, adj.P.Val) %>% 
  mutate(genotype_direction = paste0(genotype, " ",  direction)) %>% 
  mutate(genotype_direction = factor(genotype_direction, levels = c("ATRX up", "ATRX down", "TERT up", "TERT down")))

# ----
# Organise the bulk gene expression matrix to plot 
# ----

# Z score transform
lcpm_scaled <- t(scale(t(lcpm_list$NoHN_Core)))

# subset the gene expression matrix to just include the genes i want to plot
bulk_heatmap_deg <- lcpm_scaled[genes_plot_df$Gene, ]
# make genes rownames
rownames(bulk_heatmap_deg) <- str_extract(rownames(bulk_heatmap_deg),
                                          pattern = '[^_]*$')

# check that the order of the heatmap columns hasn't changed 
table(a5_anno.core$A5_ID == colnames(lcpm_scaled))

# ----
# Draw the bulk heatmap
# ----

# specify the colour gradient to use for this heatmap
col_fun = colorRamp2(
  seq(min(bulk_heatmap_deg), max(bulk_heatmap_deg), len = 11),
  rev(brewer.pal(11, "RdYlBu")))
col_fun2 = colorRamp2(
  seq(min(a5_anno.core$telhunter_log2_telcontentratio),
      max( a5_anno.core$telhunter_log2_telcontentratio), len = 9),
  brewer.pal(9, "Greens"))

genotype_cols <- setNames(c(pal_d3()(2), "lightgrey"),
                          nm=unique(a5_anno.core$genotype_groups))

# make an annotation describing TERT or ATRX mutation status
top_anno_bulk <- HeatmapAnnotation(
  "Genotype" = a5_anno.core$genotype_groups,
  "Telomere length ratio" = as.numeric(a5_anno.core$telhunter_log2_telcontentratio), 
  "Sex" = a5_anno.core$Gender, 
  show_annotation_name = c("Genotype" = FALSE,
                           "Telomere length ratio" = FALSE,
                           "Sex" = FALSE),
  annotation_legend_param = list(
    "Genotype" = list(legend_height = unit(3, "cm")),
    "Telomere length ratio" = list(legend_height = unit(3, "cm"))
    ),
  col = list("Genotype" = genotype_cols,
             "Telomere length ratio" = col_fun2,
             "Sex" = gender_cols),
  show_legend = c("Genotype" = TRUE)
  )

# plot the heatmap
bulk_hm <- Heatmap(bulk_heatmap_deg,
                   #col = col_fun,
                   row_split = genes_plot_df$genotype_direction,
                   column_split = a5_anno.core$genotype_groups,
                   width = ncol(bulk_heatmap_deg)*unit(2, "mm"),
                   height = ncol(bulk_heatmap_deg)*unit(4, "mm"),
                   row_gap = unit(2, "mm"),
                   column_gap = unit(2, "mm"),
                   row_title_rot = 0,
                   top_annotation = top_anno_bulk,
                   column_title_rot = 90,
                   cluster_columns = FALSE,
                   #show_column_dend = FALSE,
                   cluster_column_slices = FALSE,
                   cluster_rows = FALSE,
                   #show_row_dend = FALSE,
                   cluster_row_slices = FALSE,
                   show_row_names = TRUE,
                   row_names_side = "left",
                   show_column_names  = FALSE,
                   row_names_gp = gpar(fontface = "italic"),
                   heatmap_legend_param  = hm_legend_params)
bulk_hm

```

```{r gsea}
collections <- list.files("/data/cephfs/punim0648/Pattison_projects/A5/A5-paper/RNA-Seq/Reference data/MSIGdb/", full.names = T,pattern = "*.gmt")

# Save each toptable and run GSEA analysis on each contrast
for(i in 1:length(colnames(contr.matrix))){
  
  contrast <- colnames(contr.matrix)[i]
  toptable <- topTable(efit,coef=contrast,sort.by="p",number = Inf) %>% 
    tibble::rownames_to_column("EnsGID_Symbol") %>% 
    separate(EnsGID_Symbol, into=c("EnsGID", "Symbol"),extra = "merge", sep="_")
  
  # Get the top 20 symbols to plot
  top_FC <- toptable %>%
    filter(adj.P.Val < 0.05) %>%
    filter(!is.na(Symbol)) %>%
    arrange(-abs(logFC))
  
  top_p <- toptable %>%
    filter(adj.P.Val < 0.05)%>%
    arrange(adj.P.Val)%>%
    filter(!is.na(Symbol))
  
  keep_genes <- c(top_FC$Symbol[1:20],top_p$Symbol[1:20])
  keep_genes<- keep_genes[!is.na(keep_genes)]
  
  
  plot_input <- toptable %>%
    mutate(label = paste0("italic('",Symbol, "')"))%>%
    mutate(colour = ifelse(adj.P.Val <= 0.05, "FDR <= 0.05", "FDR > 0.05"))%>%
    mutate(FC = ifelse(sign(logFC) > 0, "Up", "Down"))%>%
    mutate(colour = paste(colour, FC))%>%
    mutate(colour = replace(colour, colour %in% c("FDR > 0.05 Up", "FDR > 0.05 Down"),"FDR > 0.05"))%>%
    mutate(colour = factor(colour, levels = c("FDR > 0.05","FDR <= 0.05 Down","FDR <= 0.05 Up")))%>%
    arrange(adj.P.Val)%>%
    mutate(label = replace(label,!Symbol %in% keep_genes, NA))%>%
    mutate(label = factor(label, levels = unique(label)))%>%
    arrange(colour)
  
 
  
  # MA plot
  MA <- ggplot(data = plot_input, aes (x  = AveExpr, y = logFC, colour = colour))+
    geom_point(size = 0.5, alpha = 0.8)+
    geom_text_repel(parse = T, size =5, colour ="black", aes(x  = AveExpr, y = logFC,label = label))+
    scale_colour_manual(values = c("grey","blue", "red"))+
    labs(x = expression('Mean expression (log'[2]*'CPM)'), y = expression('Log'[2]*'FC'), colour = "Significance")+
    geom_hline(yintercept = 0)+
    guides(colour = F)+
    ggsave(paste0("./results/plots/ma_plots/", contrast, "_MA.pdf"),useDingbats= F)
    
  for(collection in collections){
    
    collection_name <- gsub(".symbols.gmt","", basename(collection))
    
    gene_set <- GSA.read.gmt(collection)
    
    gene_set_formatted <- gene_set$genesets
    
    names(gene_set_formatted) <- gene_set$geneset.names
    
    indexed <- ids2indices(gene.sets = gene_set_formatted, identifiers = gsub("ENSG[0-9.]+_","", rownames(v$E)), remove.empty=TRUE)
    
    camera_result <- camera(y = v ,index = indexed, design = design_mat, contrast = contr.matrix[,contrast])%>%
      rownames_to_column("Gene set")%>%
      dplyr::select(`Gene set`,"NGenes" ,"Direction","PValue","FDR" )%>%
      filter(FDR <= 0.05)%>%
      mutate(Contrast= contrast)%>%
      write_csv(paste0("./results/differential_gene_expression/bulk_RNA_seq/gsea/Camera_",collection_name, "_", contrast,".csv"))

    fry_result_hallmark  <- fry(y = v ,index = indexed, design = design_mat, contrast = contr.matrix[,contrast])%>%
      rownames_to_column("Gene set")%>%
      dplyr::select(`Gene set`,"NGenes" ,"Direction","PValue","FDR")%>%
      filter(FDR <= 0.05)%>%
      mutate(Contrast= contrast)%>%
      write_csv(paste0("./results/differential_gene_expression/bulk_RNA_seq/gsea/Fry_",collection_name, "_", contrast,".csv"))
  }
}
```




```{r clusterProfiler}

#Human

mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl",host = "http://uswest.ensembl.org/")) #http://asia.ensembl.org

universe_background <- select(org.Hs.eg.db,keys(org.Hs.eg.db), columns=c("ENSEMBL","ENTREZID")) %>% 
  filter(!is.na(ENSEMBL)) %>% 
  pull(ENSEMBL) %>% 
  unique()

ego <- list()
for (tt in c("atrx_tt", "tert_tt", "tert_vs_atrx_tt"))
{
 ego[[tt]] <- list()
   
  toptable <- get(tt)
gene_list <- toptable %>% 
  filter(adj.P.Val < 0.05, abs(logFC)> 2) %>% 
  separate(Gene, into=c("EnsGID", "Symbol"),extra = "merge", sep="_") %>% 
  dplyr::select(EnsGID, Symbol) %>% 
  mutate(EnsGID=gsub("[.][0-9]{1,2}$","",EnsGID)) %>% 
  distinct()

for (Ont in c("CC","MF","BP"))
{
ego[[tt]][[Ont]] <- enrichGO(gene          = as.character(gene_list$EnsGID),
                universe      = universe_background,
                OrgDb         = org.Hs.eg.db,
                keyType       = 'ENSEMBL',
                ont           = Ont,
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
        readable      = TRUE)

}



# EnsGIDtoEntrez <- getBM(filters= "ensembl_gene_id", 
#                                  attributes= c("ensembl_gene_id",
#                                                "entrezgene_id",
#                                                "external_gene_name"),
#                                  values=gene_list$EnsGID,mart= mart) %>% 
#   filter(!is.na(entrezgene_id))



}

```
```{r environment}
sessionInfo()
```
