---
title: "A5 Paired Sample Analysis"
author: "Aidan Flynn"
date: "19/05/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/g/data/pq08/projects/ppgl/a5")

basedir <- "/g/data/pq08/projects/ppgl/a5"

```

Code for comparing the genomic features (variants, CNA, SVs) of paired samples in the A5 cohort


```{r source_dependencies, message=FALSE, warning=FALSE}
source(paste0(basedir,"/wgs/scripts/utility/compare_vcf_calls.r"))
source(paste0(basedir,"/wgs/scripts/data_loaders/wgs_dataloaders.r"))
source(paste0(basedir,"/sample_annotation/scripts/data_loaders/a5_clinical_annotation_dataloader.r"))
library(patchwork)
library(ggplot2)
library(googlesheets4)
library(furrr)

data_loader_a5_clinical_anno(google_account = "aidan.flynn@umccr-radio-lab.page", use_cache = T)


```
## Definitions


```{r source_definitions}

#Color Pallete
ColorPalette <- c(LightBlue1="#cadae8ff",LightBlue2="#abc4e2ff",LightBlue3="#8aa9d6ff",
                  DarkBlue1="#7884baff",DarkBlue2="#606d9fff",DarkBlue3="#4c5a88ff",
                  Purple1="#765b92ff",Purple2="#9370a5ff",Purple3="#b280a9ff",Purple4="#cc85b1ff",
                  LightRed1="#ee7474ff",LightRed2="#e2696aff",LightRed3="#de5353ff",
                  DarkRed1="#c25858ff",DarkRed2="#c04241ff",DarkOrange1="#c65a44ff",
                  DarkOrange2="#eb7b54ff",LightOrange1="#f18d73ff",LightOrange2="#f5a697ff",
                  LightBrown1="#f5b9b0ff",LightBrown2="#d6a097ff",DarkBrown1="#c17963ff",DarkBrown2="#96665aff",
                  DarkGreen1="#637b62ff",DarkGreen2="#6ea668ff",LightGreen1="#98ca8cff",LightGreen2="#e2eab5ff",
                  Yellow1="#fbf2adff",Yellow2="#fef8c6ff",Yellow3="#f4e764ff",Salmon="#f2c5a7ff",LightSalmon="#f2dec4ff",
                  LemonGrey1="#f7f0e2ff",LemonWhite1="#fefbe5ff",LemonWhite2="#f7f4dcff",LemonGrey2="#efecdcff",
                  LightGrey1="#e2e0d9ff",LightGrey2="#d6d6d4ff",DarkGrey1="#b4b4b4ff",DarkGrey2="#9a9999ff")

#Possible consequence values deleterious variants
deleterious_coding_consequences <- c(                  
  "splice_acceptor_variant",
  "splice_region_variant",
  "splice_donor_variant",
  "stop_gained",
  "stop_lost",
  "start_lost",
  "missense_variant",
  "frameshift_variant",
  "inframe_insertion",
  "inframe_deletion")

deleterious_coding_consequences.regex <- paste(deleterious_coding_consequences, collapse ="|")

#Classes of CNA events
cn_event_types <- c("Diploid/Haploid-X", "Loss",  "Subclonal Loss", "Minor Subclonal Loss",  
  "Loss + Subclonal CNLOH", "CNLOH", "Gain", "Subclonal Gain", "Gain+LOH", "WGD", "WGD+Gain",
  "Hom. Del.", "Other", "Chromothripsis")

#Genes of interest for annotation
GOI <- read.delim("/g/data/pq08/reference/gene_lists/pcawg_mutational_drivers.txt") %>% pull(Gene) %>% as.character()
GOI.E167 <- data.frame(gene=c("NRAS", "TERT", "MLH1"),
              position=c(114716126, 1295113, 37050485))

#List of A5 patients with paired samples
Samples <- list(
  E122=c("E122-T01","E122-T02"),
  E128=c("E128-T01","E128-T02"),
  E132=c("E132-T01","E132-T02"),
  E143=c("E143-T01","E143-T02","E143-T03"),
  E146=c("E146-T01","E146-T02"),
  E158=c("E158-T01","E158-T02"),
  E159=c("E159-T01","E159-T02","E159-T03","E159-T04"),
  E166=c("E166-T01","E166-T02"),
  E167=c("E167-T01","E167-T02"),
  E169=c("E169-T01","E169-T02"),
  E225=c("E225-T01","E225-T02"),
  E229=c("E229-T01","E229-T02")
)

```
## Match Variants between VCFs
```{r vcf_matching, eval=TRUE, message=FALSE, warning=FALSE, include=TRUE, cache=TRUE, cache.path="/g/data/pq08/projects/ppgl/a5/wgs/scripts/paired_sample_analysis/markdown_caches/"}
threads=6
plan(strategy = "multisession", workers=threads)

#Pairwise matching of VCFs - returns a list of dataframes with variant calls marked present/absent for each sample from each patient


matched_variant_tables <- 
  furrr::future_map(.x = Samples, 
                    .f = function(patient_samples, delregex) {
  patient = stringr::str_extract(patient_samples[[1]], "E...")
  message("Started variant matching for: ", patient)
  
  vcf_files <- paste0(basedir, "/wgs/analysis/bcbio/",
                      patient,
                      "/umccrised/",
                      gsub("T0","",patient_samples),
                      "__",
                      patient_samples,
                      "/small_variants/",
                      gsub("T0","",patient_samples),
                      "__",
                      patient_samples,"-somatic-PASS.vcf.gz")
  names(vcf_files) <- patient_samples
  
  #Swap UMCCRised VCF (incomplete due to high TMB in E167-T01) with BCBio VCF
  if(patient == "E167")
  {
    vcf_files[["E167-T01"]] <-
      paste0(basedir,"/wgs/analysis/bcbio/E167/final/2021-10-28_E167/E167-1-ensemble-annotated.vcf.gz")
    vcf_files[["E167-T02"]] <-
      paste0(basedir,"/wgs/analysis/bcbio/E167/final/2021-10-28_E167/E167-2-ensemble-annotated.vcf.gz")
  }
  
 
  return_table <-
    multi_compare_small_variant_vcf(
      vcfs = vcf_files,
      normal_names = gsub("T.+", "B01", patient_samples),
      tumour_names = patient_samples,
      genome = "BSgenome.Hsapiens.UCSC.hg38",
      Keep_Only_Highest_Consequence = T,
      collapse_annotation = T)
    
  
  
  
  # #Override missed TERT mutation due to UMCCRISE filtering it out
  # if (patient == "E167") {
  #   temp <- return_table
  #   temp[temp$seqnames == "chr5" &
  #          temp$start == 1295113, "E167.T01_Called"] <- 'Y'
  #   return_table <- temp
  #   rm(temp)
  # }
  message("Completed variant matching for: ", patient)
  
  return(return_table)
}, delregex=deleterious_coding_consequences.regex)

#Fix and samples names that have dashes replaced with dots
matched_variant_tables <-
  lapply(matched_variant_tables, function (table) {
    return(rename_with(
      table,
      .fn = ~ gsub("(E[0-9]{3}).([TB][0-9]{2})", "\\1-\\2", .x),
      .cols = matches("_Called$|_VAF$")
    ))
  })

#Annotate black listed variants
blacklist="/g/data/pq08/projects/ppgl/a5/wgs/analysis/mpileup/blacklist/blacklists/blacklist_readsupport_gteq3_samplesupport_gteq3.tsv"
blacklisted_variants <- read.delim(blacklist)

matched_variant_tables <-
  lapply(matched_variant_tables, function (table, black_list = blacklisted_variants) {
        table <- table %>% 
      left_join(black_list %>% 
                  dplyr::rename("blacklist_n_normal_observed"=n_above_threshold) %>% 
                  mutate(blacklist=TRUE),
                by = c("seqnames"="CHROM", "start"="POS", "ALT"="ALT"))
    
    table$blacklist_n_normal_observed[is.na(table$blacklist_n_normal_observed)] <- 0
    table$blacklist[is.na(table$blacklist)] <- F
    return(table %>% filter(blacklist==F))
  })

```
```{r annotate_e167, include=TRUE, cache=TRUE, cache.path="/g/data/pq08/projects/ppgl/a5/wgs/scripts/paired_sample_analysis/markdown_caches/"}

plan(strategy = "sequential")
  
#####
# Merge the annotation from the UMCCRised vcf (cancer genes only)
#####

# extract_gene_info <- function(annotation_string, delregex) {
#       symbol_consequence <- annotation_string %>% 
#                stringr::str_split(pattern=",", simplify = T) %>%
#                purrr::map(stringr::str_split, pattern="[|]", simplify=T) %>%
#                purrr::map(.f = ~"["(.x,c(2,4)), simplify=T) 
#       
#       symbol_consequence <- data.frame(symbol = sapply(symbol_consequence, "[[",2),
#                                        consequence = sapply(symbol_consequence, "[[",1))
#       symbol_consequence <-
#         symbol_consequence[grepl(x=symbol_consequence$consequence, 
#                                  pattern =  delregex),]
#               
#       return(paste(unique(symbol_consequence$symbol), collapse="|"))
#     }

#UMCCRised VCFs
vcf_files <- paste0(basedir, "/wgs/analysis/bcbio/",
                      "E167",
                      "/umccrised/",
                      gsub("T0","",Samples[["E167"]]),
                      "__",
                      Samples[["E167"]],
                      "/small_variants/",
                      gsub("T0","",Samples[["E167"]]),
                      "__",
                      Samples[["E167"]],"-somatic-PASS.vcf.gz")
  names(vcf_files) <- Samples[["E167"]]

  #Read in VCFs and extract annotation information
  E167_umccrised_data <- purrr::map(vcf_files, readVcf) %>% 
   purrr::map(function(vcf_data) {
  data.frame(seqnames=seqnames(vcf_data), 
             start=start(vcf_data),
             end=end(vcf_data),
             info(vcf_data)[c("PCGR_SYMBOL","PCGR_TIER","PCGR_CONSEQUENCE")])
  }) %>% 
    bind_rows() %>% 
    distinct()
  
  #Annotate matched data with UMCCRised subset of VCFs (cancer genes only due to high TMB)
  matched_variant_tables$E167 <- 
    matched_variant_tables$E167 %>% 
    left_join(E167_umccrised_data)

  
  ##Guess at annotation for variants not in UMCCRise
  matched_variant_tables$E167 <- matched_variant_tables$E167 %>% 
    distinct() %>% 
    group_by(seqnames) %>%  
    group_split() %>% 
    furrr::future_map(.f = function(chr_data) {
    chr_data %>% rowwise() %>% mutate(PCGR_TIER=ifelse(
    any(stringr::str_detect(pattern = deleterious_coding_consequences, string=Annotation)), 
    "TIER_4", 
    "NONCODING")) }) %>% bind_rows()
      
  
```

```{r annotate_pileups, eval=TRUE, include=TRUE}

source(file.path(basedir,"wgs/scripts/paired_sample_analysis/process_paired_pileups.r"))

if (!all(names(pileup_vafs) == names(matched_variant_tables))) { stop("Pileup and matched table names do not match") }

#annotate matched VCFs with pileup data
matched_variant_tables <- map2(.x = matched_variant_tables, 
     .y = pileup_vafs, 
     .f = \(match_table, pileup_table) left_join(match_table, 
                                                 pileup_table, 
                                                 by=c("seqnames"="CHROM", 
                                                      "start"="POS",  
                                                      "ALT"="ALT")))

#Update called status of variants using pileup data and filter variants with any support in the normal
matched_variant_tables <- map(.x = matched_variant_tables, 
     .f = function(match_table) {
       
       called_columns = colnames(match_table)[grepl("_Called",colnames(match_table))]
       tumour_samples <- gsub("_Called","",called_columns)
       normal <- gsub("-T0.","-B01",tumour_samples[1])
       
       match_table <- match_table %>% filter(is.na(!!sym(paste0(normal,"_pileup_vaf"))) |
                                               (!!sym(paste0(normal,"_pileup_vaf")) == 0 &
                                                  !!sym(paste0(normal,"_VAF")) == 0)) 
       
       for (tumour in tumour_samples){
         match_table <- match_table %>% 
           mutate("{tumour}_Called" := ifelse(!is.na(!!sym(paste0(tumour,"_pileup_vaf"))) &
                                                !!sym(paste0(tumour,"_pileup_vaf")) > 0, 
                                                            yes="Y", 
                                                            no= !!sym(paste0(tumour,"_Called")))) %>%
           #update VAF from pileup if variant missed
           mutate("{tumour}_VAF" := ifelse(!!sym(paste0(tumour,"_Called")) == "Y" &
                                             !!sym(paste0(tumour,"_VAF")) == 0 &
                                              !is.na(!!sym(paste0(tumour,"_pileup_vaf"))), 
                                                            yes=!!sym(paste0(tumour,"_pileup_vaf")), 
                                                            no= !!sym(paste0(tumour,"_VAF"))))
       }
       
       return(match_table)
       
     })

```
## Preprocess data for input to ggplots
```{r figure_input_preprocessing, eval=TRUE, include=TRUE}

plan(strategy = "multisession", workers=threads)

#Result lists
comparison_grids <- list()
dotplot_data <- list()

for (patient in names(Samples))
{
  #####################
  ## Comparison Grid ##
  #####################
  #nSample by nSample - intersection show count of common variants
  
  pairs <- make_pairs(Samples[[patient]]) #make unique combinations
  pairs <- rbind(pairs, cbind(Sample1=Samples[[patient]],Sample2=Samples[[patient]])) #add self pairings
  pairs$nShared <- 0
  
  #count events present in specified pairings
  for(i in 1:nrow(pairs))
  {
    pairs$nShared[[i]] <- matched_variant_tables[[patient]] %>% ungroup() %>% 
      filter(!!sym(paste(pairs[i,1],"Called", sep="_")) == "Y", 
             !!sym(paste(pairs[i,2],"Called", sep="_")) == "Y") %>% 
      dplyr::count() %>% pull(n)
  }
  
  #Make reciprocal pairs to fill in grid and sort/factorise to ensure order
  pairs <- unique(rbind(pairs, data.frame(Sample1=pairs$Sample2,Sample2=pairs$Sample1, nShared=pairs$nShared)))
  pairs$Sample1 <- as.character(pairs$Sample1)
  pairs$Sample2 <- as.character(pairs$Sample2)
  pairs$Sample1 <- factor(pairs$Sample1, levels = sort(unique(pairs$Sample1)))
  pairs$Sample2 <- factor(pairs$Sample2, levels = sort(unique(pairs$Sample2)))
  
  comparison_grids[[patient]] <- pairs
  
  ######################################
  ## Variant Membership Dot Plot Data ##
  ######################################
  
  
  matched_variant.longform <- matched_variant_tables[[patient]] %>% 
    ungroup() %>% 
    dplyr::select(PCGR_SYMBOL, seqnames, start, end, REF, ALT,PCGR_TIER, matches(c("E...-T0._Called", "E...-T0._VAF"))) %>% 
    mutate(across(.cols = ends_with("_Called"), .fns=~ifelse(.x=="Y", 1,0))) %>% #convert called to numeric so it rbinds with VAF
    #Pivot into long form separating  sample name and data-type into columns 
    pivot_longer(cols = matches(c("_Called$", "T[0-9]{2}_VAF$")), 
                 names_sep = "_",
                 names_to = c("A5_ID", "DataType"), 
                 values_to = "Value") %>% 
    #Pivot wider to unmix datatypes but keep samples in longform
    pivot_wider(names_from = DataType, values_from = Value)  %>% 
    #Filter to only call events
    filter(Called==1) %>% 
    group_by(seqnames) %>% group_split() %>% 
    furrr::future_map(.f=function(chr_data){
     chr_data %>% group_by(seqnames, start, end, REF, ALT) %>% 
    mutate(Privacy=paste(A5_ID,sep = "/", collapse = "/"),
           nSamplesCalled=n()) %>%  ungroup()}) %>% bind_rows()
  
  
  
  # dotplot_data[[patient]] <- matched_variant.longform  %>% 
  #   ## Create annotation groups based on variant tiers and genes of interest list 
  #   ## E167 is treated separately due to extreme mutation load
  #   mutate(Annotation=case_when((is.na(PCGR_TIER) | PCGR_TIER == "NONCODING") ~ "Non-Coding/Syn.",
  #                               PCGR_TIER %in% c("TIER_1","TIER_2","TIER_3") ~ PCGR_SYMBOL,
  #                               PCGR_TIER == "TIER_4" & PCGR_SYMBOL %in% GOI ~ PCGR_SYMBOL,
  #                               PCGR_TIER == "TIER_4" ~ "TIER_4"),
  #          #handle E167-1 high TMB special case
  #          Annotation=case_when(A5_ID == "E167-T01" & PCGR_TIER != "NONCODING" & PCGR_SYMBOL %in% GOI.E167 ~ PCGR_SYMBOL,
  #                               #Keep annotation for variants also present in E167-2
  #                               A5_ID == "E167-T01" & 
  #                                 (paste(seqnames,start) %in% (
  #                                   matched_variant.longform %>% filter(A5_ID=="E167-T02") %>% 
  #                                     transmute(coord=paste(seqnames,start)) %>% 
  #                                     pull(coord))) ~ Annotation,
  #                               A5_ID == "E167-T01" & Annotation=="Non-Coding/Syn." ~ Annotation,
  #                               A5_ID == "E167-T01" ~ PCGR_TIER,
  #                               TRUE ~ Annotation),
  #          AnnotationClass=ifelse(Annotation %in% c("TIER_1", "TIER_2", "TIER_3", "TIER_4","Non-Coding/Syn."),"Summary","Specific"),
  #          AnnotationClass=factor(as.character(AnnotationClass), levels = c("Specific","Summary")),
  #          VAF=ifelse(AnnotationClass=="Specific",VAF,NA),
  #          Position=ifelse(AnnotationClass=="Specific",start,"")) %>% 
  #   #Keep only one representative entry for each annotation group and annotate with variant count
  #   group_by(Annotation, Position, A5_ID, Privacy) %>% mutate(variant_group_size=n()) %>% slice_head(n=1) %>% 
  #   arrange(desc(AnnotationClass), desc(nSamplesCalled), Privacy) %>% 
  #   mutate(Annotation=ifelse(AnnotationClass == "Summary", paste0(Annotation," (n=",variant_group_size,")"), Annotation)) %>%
  #   mutate(Annotation=gsub("TIER_", "Tier ", Annotation)) %>% 
  #   mutate(Annotation=factor(Annotation, levels = unique(.$Annotation))) %>% 
  #   mutate(LabelA=paste(Annotation, Privacy, sep="_"), LabelA=factor(LabelA, levels = unique(LabelA)))
  
  dotplot_data[[patient]] <- matched_variant.longform  %>% 
    mutate(PCGR_SYMBOL=replace_na(PCGR_SYMBOL, "")) %>% 
    ## Create annotation groups based on variant tiers and genes of interest list
    ## E167 is treated separately due to extreme mutation load
    mutate(Annotation=case_when((
      is.na(PCGR_TIER) | PCGR_TIER == "NONCODING") ~ "Non-Coding/Syn.",
      PCGR_TIER %in% c("TIER_1","TIER_2","TIER_3") ~ PCGR_SYMBOL,
      PCGR_TIER == "TIER_4" & PCGR_SYMBOL %in% GOI ~ PCGR_SYMBOL,
      PCGR_TIER == "TIER_4" ~ "TIER_4"),
     #handle E167-1 high TMB special case
      Annotation=case_when(
        A5_ID %in% c("E167-T01","E167-T02") & (PCGR_SYMBOL %in% GOI.E167$gene & start %in% GOI.E167$position) ~ PCGR_SYMBOL,
        #Keep annotation for variants also present in E167-2
        A5_ID == "E167-T01" & nSamplesCalled==2 ~ Annotation,
        A5_ID == "E167-T01" & Annotation=="Non-Coding/Syn." ~ Annotation,
        A5_ID == "E167-T01" ~ PCGR_TIER,
        TRUE ~ Annotation),
      AnnotationClass=ifelse(Annotation %in% c("TIER_1", "TIER_2", 
                                               "TIER_3", "TIER_4",
                                               "Non-Coding/Syn."),"Summary","Specific"),
           AnnotationClass=factor(as.character(AnnotationClass), levels = c("Specific","Summary")),
           VAF=ifelse(AnnotationClass=="Specific",VAF,NA),
           Position=ifelse(AnnotationClass=="Specific",start,""))
 
  #Reduce annotation Specificity to just coding/noncoding - optional
  dotplot_data[[patient]] <- dotplot_data[[patient]] %>%  mutate(Annotation=gsub("TIER_.", "Coding", Annotation))
   
 dotplot_data[[patient]] <- dotplot_data[[patient]] %>%
    #Keep only one representative entry for each annotation group and annotate with variant count
    group_by(Annotation, Position, A5_ID, Privacy) %>% mutate(variant_group_size=n()) %>% slice_head(n=1) %>%
    arrange(desc(AnnotationClass), desc(nSamplesCalled), Privacy) %>%
    mutate(Annotation=ifelse(AnnotationClass == "Summary", paste0(Annotation," (n=",variant_group_size,")"), Annotation)) %>%
    mutate(Annotation=gsub("TIER_", "Tier ", Annotation)) %>%
   ungroup() %>% 
   #Handle edge cases where the same gene is mutated in multiple samples at different positions
   mutate(dupSymbol=(duplicated(PCGR_SYMBOL) | duplicated(PCGR_SYMBOL, fromLast=T)),
          dupSymPos=(duplicated(paste(PCGR_SYMBOL,Position)) | duplicated(paste(PCGR_SYMBOL,Position), fromLast=T)),
          Annotation=ifelse(AnnotationClass=="Specific" & dupSymbol & !dupSymPos, 
                             paste0(Annotation, "(g.",Position,")"),
                             Annotation)) %>% 
    mutate(Annotation=factor(Annotation, levels = unique(.$Annotation))) %>%
    mutate(LabelA=paste(Annotation, Privacy, sep="_"), LabelA=factor(LabelA, levels = unique(LabelA)))

}
```

```{r variant_plots, eval=TRUE, include=TRUE}
#VariantPlots
variant_dot_plots <- list()
variant_grid_plots <- list()
composed_plots <- list()
for (patient in names(Samples))
{
  #Recoding A5_IDs to publication IDs 
  name_recode <- 
    a5_anno %>% 
    filter(`Patient ID`==patient) %>% 
    dplyr::select(A5_ID,PublicationID) %>% 
    { set_names(x = .$PublicationID,nm = .$A5_ID) }
   name_recode_levels <- c(sort(name_recode[grepl("P",name_recode)]),
                          sort(name_recode[grepl("R",name_recode)]),
                          sort(name_recode[grepl("M",name_recode)]))
  
  comparison_grid_current <- comparison_grids[[patient]] %>% 
    mutate(across(.cols = c(Sample1,Sample2), 
                  .fns = \(x){ factor(dplyr::recode(gsub("T0","",x),
                                                    !!!name_recode), 
                                      levels=name_recode_levels)}))
  
  dotplot_current <- dotplot_data[[patient]] %>% 
    mutate(A5_ID= factor(dplyr::recode(gsub("T0","",A5_ID),!!!name_recode), 
                                      levels=name_recode_levels))
  
  #Make plots
  variant_grid_plots[[patient]] <-
    ggplot(comparison_grid_current, aes(x = Sample1, y = Sample2, label =
                                              nShared)) +
    geom_text(size = 3) +
    geom_hline(yintercept = seq(1.5, length(unique(
      comparison_grid_current$Sample1
    ))), color = "grey") +
    geom_vline(xintercept = seq(1.5, length(unique(
      comparison_grid_current$Sample1
    ))), color = "grey") +
    theme_bw() +
    theme(
      panel.grid = element_blank(),
      axis.title.x = element_blank(),
      #axis.text.x=element_blank(),
      axis.text.x = element_text(
        angle = 90,
        vjust = 0.5,
        hjust = 1
      ),
      #axis.ticks.x=element_blank(),
      axis.title.y = element_blank()
    ) +
    scale_y_discrete(limits = rev(levels(comparison_grid_current$Sample2))) #+
    #scale_x_discrete(position = "top")
  
  
  variant_dot_plots[[patient]] <-
    ggplot(dotplot_current, aes(x = A5_ID, y = LabelA)) +
    geom_point() +
    theme_bw() +
    theme(
      #axis.text.x = element_text(angle = 45, hjust = 1),
      axis.text.x = element_blank(),
      axis.title.y = element_blank(),
      axis.title.x = element_blank()
    )  + 
    #This is to handle when two different privacy states have the same Annotation, 
    #using a combined privacy/annotation label prevents them colliding,
    #the combined label is then overwritten by the annotation alone
  scale_y_discrete(labels=dotplot_current %>% ungroup() %>% dplyr::select(LabelA, Annotation) %>% distinct() %>% pull(Annotation))

  ##Add gradient coloring when VAF info is present
  # if (!all(is.na(dotplot_data[[patient]]$VAF))) {
  #   p2 <- p2 + aes(color=VAF) + 
  #     scale_color_gradientn(values = seq(0,1,0.2),
  #                           colours = c("#4d39b5ff",
  #                                       "#39b4b5ff",
  #                                       "#40b539ff",
  #                                       "#b5a739ff",
  #                                       "#b54b39ff")
  #                           , na.value = "black", limits=c(0,1))
  # }
  
  composed_plots[[patient]] <-
    variant_dot_plots[[patient]] + theme(plot.margin = margin(6, 6, 0, 6)) +
    variant_grid_plots[[patient]] +
    plot_layout(nrow = 2, heights = c(1, 0.3))
}


```

```{r write_variant_dot_plots, echo=F, eval=FALSE, include=FALSE}

  lapply(names(composed_plots), function(patient, composed_plots, dotplot_data){
    ggsave(filename = paste0("./wgs/results/paired_sample_analysis/",patient,"_variant_only.pdf"),
           plot = composed_plots[[patient]],
           device = pdf(),
           width = 3+(0.3*length(unique(dotplot_data[[patient]]$A5_ID))),
           height=3+(0.07*nrow(dotplot_data[[patient]])),
           limitsize = F)
    dev.off()
  }, composed_plots, dotplot_data) 

```

# CNA Data Loading
```{r load_cna_data}

A5_seg.files <- list.files(paste0(basedir,"/wgs/symlinks/purple_seg"), pattern = ".purple.segment.tsv", full.names = T ,recursive = T)
names(A5_seg.files) <- gsub(".purple.segment.tsv", "", basename(A5_seg.files))
A5_seg <- lapply(A5_seg.files, read.delim)

A5_seg <- bind_rows(A5_seg, .id="A5_ID")
A5_seg <- A5_seg %>% mutate(A5_ID=gsub("-T0", "-", A5_ID))
A5_seg <- A5_seg %>%  dplyr::filter(chromosome!="chrY")

A5_seg.keep <- A5_seg  %>%  filter(germlineStatus != "NOISE" & 
                                     (
                                       (end-start > 1000000 & (bafCount > 20 | chromosome=="chrX") & depthWindowCount > 100) | 
                                        support %in% c("BND","DEL","SGL","DUP","INV","MULTIPLE","INF")
                                      )
                                   )

A5_seg.keep <- A5_seg.keep %>% left_join(a5_anno %>%  dplyr::select(A5_ID, Gender))
A5_seg.keep <- A5_seg.keep %>% group_by(A5_ID) %>% mutate(mean_tumorCopyNumber=mean(tumorCopyNumber)) %>%  ungroup()

chr_offsets <- A5_seg.keep %>% 
  mutate(chromosome=factor(as.character(chromosome), levels=paste0("chr",c(1:22,"X")))) %>% 
  arrange(chromosome) %>% group_by(A5_ID, chromosome) %>% 
  summarise(max_end=max(end)) %>% 
  ungroup() %>% group_by(chromosome) %>% 
  summarise(chr_size=median(max_end)) %>% mutate(offset = cumsum(as.numeric(chr_size))-chr_size)

A5_seg.keep <- A5_seg.keep %>% left_join(chr_offsets) %>% mutate(start_offset=start+offset, end_offset=end+offset)

A5_seg.keep <- A5_seg.keep %>% mutate(
  Class = case_when(
    (majorAlleleCopyNumber >= 1.85 &
       majorAlleleCopyNumber <= 2.15) &
      (minorAlleleCopyNumber >= 1.85 &
         minorAlleleCopyNumber <= 2.15) &
      (mean_tumorCopyNumber >= 3) ~ "WGD",
    (majorAlleleCopyNumber >= 2.85) &
      (minorAlleleCopyNumber >= 1.85 &
         minorAlleleCopyNumber <= 2.15) &
      (mean_tumorCopyNumber >= 3) ~ "WGD+Gain",
    (majorAlleleCopyNumber >= 0.5 &
       majorAlleleCopyNumber <= 1.0) &
      (minorAlleleCopyNumber >= 0.97 &
         minorAlleleCopyNumber <= 1.15) ~ "Diploid/Haploid-X",
    (majorAlleleCopyNumber >= 0.5 &
       majorAlleleCopyNumber <= 1.15) &
      (minorAlleleCopyNumber <= 0.25) &
      (chromosome != "chrX" |
         (chromosome == "chrX" & Gender == "female")) ~ "Loss",
    (majorAlleleCopyNumber >= 1.15 &
       majorAlleleCopyNumber <= 1.85) &
      (minorAlleleCopyNumber <= 0.25) &
      (chromosome != "chrX" |
         (chromosome == "chrX" &
            Gender == "female")) ~ "Loss + Subclonal CNLOH",
    (majorAlleleCopyNumber >= 0.5 &
       majorAlleleCopyNumber <= 1.15) &
      (minorAlleleCopyNumber <= 0.25) &
      (chromosome == "chrX" &
         Gender == "male") ~ "Diploid/Haploid-X",
    (majorAlleleCopyNumber >= 0.5 &
       majorAlleleCopyNumber <= 1.15) &
      (minorAlleleCopyNumber <= 0.85) &
      (chromosome != "chrX" |
         (chromosome == "chrX" &
            Gender == "female")) ~ "Subclonal Loss",
    (majorAlleleCopyNumber >= 0.5 &
       majorAlleleCopyNumber <= 1.15) &
      (minorAlleleCopyNumber <= 0.97) &
      (chromosome != "chrX" |
         (chromosome == "chrX" &
            Gender == "female")) ~ "Minor Subclonal Loss",
    (majorAlleleCopyNumber >= 1.85 &
       majorAlleleCopyNumber <= 2.15) &
      (minorAlleleCopyNumber <= 0.25) ~ "CNLOH",
    (majorAlleleCopyNumber >= 1.85) &
      (minorAlleleCopyNumber >= 0.85) ~ "Gain",
    #& minorAlleleCopyNumber <= 1.15
    (majorAlleleCopyNumber >= 1.0) &
      (minorAlleleCopyNumber >= 0.85) ~ "Subclonal Gain",
    #& minorAlleleCopyNumber <= 1.15
    (majorAlleleCopyNumber >= 1.85) &
      (minorAlleleCopyNumber <= 0.85) ~ "Gain+LOH",
    tumorCopyNumber <= 0.5 ~ "Hom. Del.",
    TRUE ~ "Other"
  )
) %>%
  mutate(
    Class = ifelse((
      Class == "Minor Subclonal Loss" &
        (
          !is.na(lead(Class)) & lead(Class) == "Subclonal Loss"
        ) |
        (!is.na(lag(Class))  &
           lag(Class) == "Subclonal Loss")
    ),
    "Subclonal Loss",
    Class),
    Class = ifelse((
      Class == "Minor Subclonal Loss" &
        (
          !is.na(lead(Class)) & lead(Class) == "Diploid/Haploid-X"
        ) |
        (!is.na(lag(Class))  &
           lag(Class) == "Diploid/Haploid-X")
    ),
    "Diploid/Haploid-X",
    Class
    )
  ) %>%
  mutate(Class = factor(Class, levels = cn_event_types))

#Manual overide for chromothripsis regions
A5_seg.keep <- A5_seg.keep %>% mutate(Class=case_when(
  ((A5_ID=="E123-1" | A5_ID=="E123-T01") & chromosome=="chr5") ~ "Chromothripsis",
  ((A5_ID=="E128-2" | A5_ID=="E128-T02") & (chromosome=="chr1" & start > 124*10^6)) ~ "Chromothripsis",
  ((A5_ID=="E138-1" | A5_ID=="E138-T01") & chromosome=="chr2") ~ "Chromothripsis",
  ((A5_ID=="E180-1" | A5_ID=="E180-T01") & (chromosome=="chr3" | chromosome=="chr11")) ~ "Chromothripsis",
  ((A5_ID=="E231-1" | A5_ID=="E231-T01") & (chromosome=="chr3" & start > 50*10^6)) ~ "Chromothripsis",
  ((A5_ID=="E198-1" | A5_ID=="E198-T01") & (chromosome=="chr17" & start < 10*10^6)) ~ "Chromothripsis",
  ((A5_ID=="E198-1" | A5_ID=="E198-T01") & (chromosome=="chr22" & ((start > 17*10^6) & (end < 50*10^6)))) ~ "Chromothripsis",
  TRUE ~ as.character(Class)
))


min_gap <-5000000
A5_seg.keep.merged <- A5_seg.keep %>% mutate(Class=gsub("Minor Subclonal Loss","Subclonal Loss", Class)) %>% arrange(A5_ID, chromosome,start,end) %>% group_by(A5_ID, chromosome) %>% 
  mutate(
    # lead_class=lead(Class),     
    # lag_class=lag(Class),
    #      lag_start=lag(start),
    #      lag_end=lag(end),
    #      diff_start_lend=start-lag(end),
    #      diff_lstart_end=lead(start) - end,
    seg_boundary=case_when(
      (Class != lead(Class) & Class != lag(Class)) ~ "Standalone",
      ((start-lag(end) > min_gap | is.na(lag(end))) & (is.na(lead(start)) | lead(start) - end > min_gap)) ~ "Standalone",
      (Class != lead(Class) & (start-lag(end) > min_gap | is.na(lag(end)))) ~ "Standalone",
      (Class != lag(Class) & (is.na(lead(start)) | lead(start) - end > min_gap)) ~ "Standalone",
      ((Class != lag(Class)) | (start-lag(end) > min_gap) | is.na(lag(start))) ~ "Begin",
      Class != lead(Class) | lead(start) - end > min_gap | is.na(lead(start)) ~ "End",
      TRUE ~ NA_character_
    )
  ) %>% 
  filter(!is.na(seg_boundary)) %>% 
  mutate(new_seg_start=ifelse(seg_boundary %in% c("Standalone", "Begin"), start, lag(start)),
         new_seg_end=ifelse(seg_boundary %in% c("Standalone", "End"), end, lead(end))) %>% 
  dplyr::select(-start, -end, -seg_boundary) %>% 
  group_by(A5_ID,chromosome,Gender,Class, new_seg_start, new_seg_end, offset) %>% slice_head(n=1) %>% ungroup() %>% 
  arrange(A5_ID,chromosome, new_seg_start, new_seg_end)

#Annotate merged segments with median of values from constituant segments
A5_seg.keep.merged$observedBAF <- NA
A5_seg.keep.merged$minorAlleleCopyNumber <- NA
A5_seg.keep.merged$majorAlleleCopyNumber <- NA
A5_seg.keep.merged$tumorCopyNumber <- NA
for (i in 1:nrow(A5_seg.keep.merged))  
{
  SOI <- A5_seg.keep[A5_seg.keep$A5_ID==A5_seg.keep.merged$A5_ID[i] &
                                               A5_seg.keep$chromosome==A5_seg.keep.merged$chromosome[i] &
                                               A5_seg.keep$start >= A5_seg.keep.merged$new_seg_start[i] &
                                               A5_seg.keep$end <= A5_seg.keep.merged$new_seg_end[i],]
  
  A5_seg.keep.merged$observedBAF[i] <- median(SOI$observedBAF) 
  A5_seg.keep.merged$minorAlleleCopyNumber[i] <- median(SOI$minorAlleleCopyNumber, na.rm = T) 
  A5_seg.keep.merged$majorAlleleCopyNumber[i] <- median(SOI$majorAlleleCopyNumber, na.rm = T)
  A5_seg.keep.merged$tumorCopyNumber[i] <- median(SOI$tumorCopyNumber, na.rm = T)
}

  
A5_seg.keep.merged <- A5_seg.keep.merged %>% 
  dplyr::rename(start=new_seg_start, end=new_seg_end) %>%  
  mutate(start_offset=start+offset, end_offset=end+offset) %>% 
  mutate(Class=factor(Class,levels=cn_event_types)) %>% 
  arrange(A5_ID, chromosome,start,end) %>% 
  mutate(Class=case_when(
  Class=="Subclonal Gain" & majorAlleleCopyNumber < 1.15 ~ factor("Diploid/Haploid-X", levels=cn_event_types),
  Class=="Minor Subclonal Loss" & minorAlleleCopyNumber > 0.95 ~ factor("Diploid/Haploid-X", levels=cn_event_types),
  TRUE ~ Class
))

A5_seg.keep <- A5_seg.keep %>%  mutate(A5_ID = gsub("-","-T0", A5_ID))
A5_seg.keep.merged <- A5_seg.keep.merged %>%  mutate(A5_ID = gsub("-","-T0", A5_ID))

```
# CNA Diagnostic
```{r cna_diagnostic, fig.height=10, fig.width=10}

cna_palette <- c(`None`=ColorPalette[["LightGrey1"]],
                  Loss=ColorPalette[["DarkRed1"]],
                  `Subclonal Loss`="#fdadadff",
                  `Hom. Del.`=ColorPalette[["Yellow1"]], CNLOH="#f97344ff", 
                  Gain=ColorPalette[["DarkBlue3"]],
                  `Subclonal Gain`=ColorPalette[["LightBlue1"]],
                  WGD=ColorPalette[["Purple2"]],
                  Chromothripsis=ColorPalette[["LightGreen1"]],
                  Other=ColorPalette[["DarkGrey2"]],
                  `Gain+LOH`=ColorPalette[["LightBlue2"]],
                  `WGD+Gain`=ColorPalette[["DarkBlue1"]], 
                  `Loss + Subclonal CNLOH`=ColorPalette[["LightOrange2"]],
                  `Diploid/Haploid-X`=ColorPalette[["LightGrey1"]],
                  `Minor Subclonal Loss`= ColorPalette[["Salmon"]])

for (s in c("E128-T02",  "E143-T01", "E159-T03", "E233-T01",  "E167-T01", "E229-T01"))
{
  gridExtra::grid.arrange(
    ggplot(
      A5_seg.keep %>%  filter(A5_ID == s),
      aes(x = start_offset, xend = end_offset, color = Class)) +
      geom_segment(aes(y = minorAlleleCopyNumber, yend = minorAlleleCopyNumber), size = 1) +
      geom_segment(aes(y = majorAlleleCopyNumber, yend = majorAlleleCopyNumber), size = 1.5) +
      geom_vline(data = chr_offsets, mapping = aes(xintercept = offset), size = 0.3) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), 
            panel.grid = element_blank()) +
      scale_x_continuous(breaks = chr_offsets$offset, labels = chr_offsets$chromosome, expand = c(0, 0)) +
      scale_color_manual(values = cna_palette) +
      ylab("Copy Number") + facet_wrap("A5_ID", ncol = 1, scale = "free_y")
    ,
    ggplot(A5_seg.keep.merged %>%  filter(A5_ID == s),
      aes(x = start_offset, xend = end_offset, color = Class)) +
      geom_segment(aes(y = minorAlleleCopyNumber, yend = minorAlleleCopyNumber), size = 1) +
      geom_segment(aes(y = majorAlleleCopyNumber, yend = majorAlleleCopyNumber), size = 1.5) +
      geom_vline(data = chr_offsets, mapping = aes(xintercept = offset), size = 0.3) +
      theme_bw() +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), panel.grid = element_blank()) +
      scale_x_continuous(breaks = chr_offsets$offset, labels = chr_offsets$chromosome, expand = c(0, 0)) +
      scale_color_manual(values = cna_palette) +
      ylab("Copy Number") + facet_wrap("A5_ID", ncol = 1, scale =
                                         "free_y")
    
  )
}
```

```{r CNA_plots}


zip <- function(...) {
  mapply(list, ..., SIMPLIFY = FALSE)
}
alternating_labels <- c(unlist(zip(paste0("chr", seq(1,20,2)),"")),"","","chrX")

cna_plots_vertical <- list()
for (patient in names(Samples))
{
  
  #Recoding A5_IDs to publication IDs 
  name_recode <- 
    a5_anno %>% 
    filter(`Patient ID`==patient) %>% 
    dplyr::select(A5_ID,PublicationID) %>% 
    { set_names(x = .$PublicationID,nm = .$A5_ID) }
  name_recode_levels <- c(sort(name_recode[grepl("P",name_recode)]),
                          sort(name_recode[grepl("R",name_recode)]),
                          sort(name_recode[grepl("M",name_recode)]))
  
cna_plots_vertical[[patient]] <- ggplot(A5_seg.keep.merged %>% 
         mutate(Class=forcats:::fct_recode(.f=Class,  
                                           Loss="Loss + Subclonal CNLOH",
                                           Gain="WGD+Gain",
                                           Gain="Gain+LOH",
                                           #None="Minor Subclonal Loss", 
                                           None="Diploid/Haploid-X")) %>% 
         filter(A5_ID %in% Samples[[patient]]) %>% 
          mutate(A5_ID=dplyr::recode(gsub("T0","",A5_ID),!!!name_recode),
                 A5_ID=factor(A5_ID, levels = name_recode_levels)), 
       aes(x=A5_ID, xend=A5_ID, y=start_offset, yend=end_offset, color=Class)) + 
  geom_segment(size=2.5) +
  geom_hline(data = chr_offsets, mapping=aes(yintercept=offset), size=0.1) +
  theme_bw() +
  theme(axis.text.x = element_text(angle=90, vjust = 0.5,hjust=1), 
        axis.ticks.y = element_blank(),
        panel.grid = element_blank()) +
  scale_y_reverse(breaks = chr_offsets$offset, labels = alternating_labels, expand=c(0,0)) +
  scale_color_manual(values=c(`None`=ColorPalette[["LightGrey1"]], Loss=ColorPalette[["DarkRed1"]], 
                              `Subclonal Loss`="#fdadadff", `Hom. Del.`=ColorPalette[["Yellow1"]], CNLOH="#f97344ff", 
                              Gain=ColorPalette[["DarkBlue3"]], `Subclonal Gain`=ColorPalette[["LightBlue1"]],
                              WGD=ColorPalette[["Purple2"]], Chromothripsis=ColorPalette[["LightGreen1"]], Other=ColorPalette[["DarkGrey2"]]
                              )) + 
  ylab("Copy Number")
}

```

## SV Data
```{r load_svs, eval=TRUE, include=TRUE}

data_loader_sv_gridsslinx(quickload = T)

sv_comparison <- list()
sv_dotplot_data <- list()
for (patient in names(Samples))
{
  pairs <- make_pairs(Samples[[patient]])
  
  pair_comps <- list()
  
  for (i in 1:nrow(pairs))
  {
    
    pair_svs <- A5_gridss_keep %>%  filter(A5_ID %in% pairs[i, ])
    
    pair_comps[[paste(pairs[i, 1], pairs[i, 2], sep = "_")]] <-
      pair_svs %>% 
      dplyr::select(A5_ID, chrom1, start1, end1, A5_ID, chrom2, start2, end2) %>% 
      distinct() %>%
      mutate(end2=replace_na(end2,0)) %>% 
      pivot_wider(id_cols = c(chrom1, start1, end1, chrom2, start2, end2), 
                  names_from = A5_ID, values_from = A5_ID) %>% 
      mutate("{pairs[i, 1]}":=ifelse( is.na(!!sym(pairs[i, 1])), F, T),
             "{pairs[i, 2]}":=ifelse( is.na(!!sym(pairs[i, 2])), F, T)) %>% 
      dplyr::rename("{paste(pairs[i, 1],'Called', sep='_')}" := pairs[i, 1]) %>% 
      dplyr::rename("{paste(pairs[i, 2],'Called', sep='_')}" := pairs[i, 2]) 
      
    
    # sv_vcfs <-
    #   paste0("/g/data/pq08/projects/ppgl/a5/wgs/analysis/purple/",
    #     pairs[i, ],"/",pairs[i, ], ".purple.sv.vcf.gz"
    #   )
    # 
    # vcf1 <- readVcf(sv_vcfs[[1]])
    # vcf1 <- vcf1[filt(vcf1)=="PASS",]
    # RP <- geno(vcf1)[["RP"]][, which(colnames(vcf1) == pairs[i, 1])]
    # SR <- geno(vcf1)[["SR"]][, which(colnames(vcf1) == pairs[i, 1])]
    # keep_vcf1 <- (RP > 7 | SR > 7 |  (RP > 4 & SR > 4))
    # keep_vcf1[is.na(keep_vcf1)] <- FALSE
    # vcf1 <- vcf1[keep_vcf1,]
    # 
    # vcf2 <- readVcf(sv_vcfs[[2]])
    # vcf2 <- vcf2[filt(vcf2)=="PASS",]
    # RP <- geno(vcf2)[["RP"]][, which(colnames(vcf2) == pairs[i, 2])]
    # SR <- geno(vcf2)[["SR"]][, which(colnames(vcf2) == pairs[i, 2])]
    # keep_vcf2 <- (RP > 7 | SR > 7 |  (RP > 4 & SR > 4))
    # keep_vcf2[is.na(keep_vcf2)] <- FALSE
    # vcf2 <- vcf2[keep_vcf2,]
    # 
    # pair_comps[[paste(pairs[i, 1], pairs[i, 2], sep = "_")]] <-
    #   compare_sv_vcfs(
    #     vcf1 = vcf1,
    #     vcf2 = vcf2,
    #     normal_name_vcf1 = paste0(patient, "-B01"),
    #     tumour_name_vcf1 = pairs[i, 1],
    #     normal_name_vcf2 = paste0(patient, "-B01"),
    #     tumour_name_vcf2 = pairs[i, 2],
    #     genome = "BSgenome.Hsapiens.UCSC.hg38",
    #     maxgap = 0,
    #     sizemargin = 0
    #   )  %>% filter(!grepl("o$", partner))
    
  }
  
  # #Fix and samples names that have dashes replaced with dots
  # pair_comps <-
  #   lapply(pair_comps, function (table) {
  #     return(rename_with(
  #       table,
  #       .fn = ~ gsub("(E[0-9]{3}).([TB][0-9]{2})", "\\1-\\2", .x),
  #       .cols = matches("_Called$|_svID$")
  #     ))
  #   })
  
  #Merge 1v1 comparisons together
  sv_comparison[[patient]] <-
    purrr::reduce(
      .x = pair_comps,
      .f = function(comp1, comp2) {
        comp1 %>%
          full_join(comp2) %>%
          mutate(across(matches("_Called"),
                        ~replace_na(.x, FALSE)))
      }
    )
  # sv_comparison[[patient]] <-
  #   purrr::reduce(
  #     .x = pair_comps,
  #     .f = function(comp1, comp2) {
  #       comp1 %>% dplyr::select(seqnames, start, end, matches("_Called")) %>%
  #         full_join(comp2 %>% dplyr::select(seqnames, start, end, matches("_Called"))) %>% 
  #         mutate(across(matches("_Called"), 
  #                       ~replace_na(.x, FALSE)))
  #     }
  #   )

  sv_dotplot_data[[patient]] <- sv_comparison[[patient]] %>% 
    group_by(across(matches("_Called"))) %>% 
    count() %>%
    unite(col="Combo", matches("_Called"), sep="/", remove=F) %>% 
    mutate(Annotation=paste0("SV (n=",n,")")) %>% 
    dplyr::select(-n) %>% 
    pivot_longer(cols = c(-Annotation,-Combo), names_to="A5_ID", values_to="Present") %>% filter(Present) %>% 
    mutate(LabelA=paste(Annotation, Combo, sep="_")) %>% 
  group_by(LabelA) %>% 
    mutate(nSamples=n()) %>% arrange(desc(nSamples)) %>% 
    mutate(LabelA=factor(LabelA, levels=unique(.$LabelA))) %>% 
    mutate(A5_ID=factor(gsub("_Called", "", A5_ID), levels=sort(Samples[[patient]])))
  
}
  
```
```{r sv_dot_plots}

sv_dot_plots <- list()

for (patient in names(Samples))
{
  
  #Recoding A5_IDs to publication IDs 
  name_recode <- 
    a5_anno %>% 
    filter(`Patient ID`==patient) %>% 
    dplyr::select(A5_ID,PublicationID) %>% 
    { set_names(x = .$PublicationID,nm = .$A5_ID) }
  
  name_recode_levels <- c(sort(name_recode[grepl("P",name_recode)]),
                          sort(name_recode[grepl("R",name_recode)]),
                          sort(name_recode[grepl("M",name_recode)]))
  
  sv_dotplot_current <- sv_dotplot_data[[patient]] %>% 
    mutate(A5_ID=factor(dplyr::recode(gsub("T0","",A5_ID), !!!name_recode), 
                        levels = name_recode_levels))
  
sv_dot_plots[[patient]] <-
    ggplot(sv_dotplot_current, aes(x = A5_ID, y = LabelA)) +
    geom_point() +
    theme_bw() +
    theme(
      #axis.text.x = element_text(angle = 45, hjust = 1),
      axis.text.x = element_blank(),
      axis.title.y = element_blank(),
      axis.title.x = element_blank()
    ) + 
  scale_x_discrete(drop=F) +
  #This is to handle when two different privacy states have the same Annotation, 
  #using a combined privacy/annotation label prevents them colliding,
  #the combined label is then overwritten by the annotation alone
  scale_y_discrete(labels=sv_dotplot_data[[patient]] %>% ungroup() %>% dplyr::select(LabelA, Annotation) %>% distinct() %>% pull(Annotation))
}

```
## Composed Plots
```{r composed_plots, eval=TRUE, fig.height=8, fig.width=5, include=TRUE}


composed_plots <- list()
for (patient in names(Samples))
{
  composed_plots[[patient]] <- 
    variant_dot_plots[[patient]] + 
    theme(plot.margin = margin(6,6,0,6), axis.ticks.length.x = unit(0.1,"cm")) + 
    variant_grid_plots[[patient]] + 
    theme(axis.text.x = element_blank(), 
          axis.ticks.length.x = unit(0.25,"cm"), 
          plot.margin = margin(0,6,0,6)) + 
    sv_dot_plots[[patient]] + 
    theme(axis.text.x.top = element_blank(), 
          axis.ticks.length.x = unit(0.25,"cm"), 
          plot.margin = margin(0,6,0,6)) + 
    cna_plots_vertical[[patient]] + 
    theme(axis.ticks.y = element_blank(), 
          plot.margin = margin(0,6,6,6)) +
    plot_layout(nrow=4, heights = c(1.3, 0.3, 0.5, 2))
  print(composed_plots[[patient]])
}


```
```{r write_composed_plots, eval=FALSE, include=FALSE}

plot_config <- 
  as_tibble(
    matrix(
      c(
        "E122", 7,5,
        "E128", 6,5,
        "E132", 6,5,
        "E143", 8,5.5,
        "E146", 5,5,
        "E158", 6,5,
        "E159", 8,5.5,
        "E166", 7,5,
        "E167", 6,5,
        "E169", 6.5,5,
        "E225", 6.5,5,
        "E229", 5,5
        ),
      byrow = T, 
      ncol = 3, 
      dimnames = list(c(),c("PatientID","height","width"))))

purrr::pwalk(.l = plot_config, 
             .f = \(PatientID, height, width) {
                    ggsave(composed_plots[[PatientID]], 
                           filename = file.path(basedir,"wgs/results/paired_sample_analysis/",
                                                paste0(PatientID,
                                                       "_pairplot_composed.pdf")
                                                ),
                           height = as.numeric(height),
                           width =  as.numeric(width),
                           units = "in")
               }
             )


```